<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_02) on Mon Nov 05 17:25:05 CET 2007 -->
<TITLE>
AbstractSupergene
</TITLE>

<META NAME="date" CONTENT="2007-11-05">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="AbstractSupergene";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
JGAP</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../../org/jgap/supergenes/Supergene.html" title="interface in org.jgap.supergenes"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html?org/jgap/supergenes/AbstractSupergene.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="AbstractSupergene.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.jgap.supergenes</FONT>
<BR>
Class AbstractSupergene</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../resources/inherit.gif" ALT="extended by "><A HREF="../../../org/jgap/BaseGene.html" title="class in org.jgap">org.jgap.BaseGene</A>
      <IMG SRC="../../../resources/inherit.gif" ALT="extended by "><B>org.jgap.supergenes.AbstractSupergene</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD>java.io.Serializable, java.lang.Comparable, <A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>, <A HREF="../../../org/jgap/IBusinessKey.html" title="interface in org.jgap">IBusinessKey</A>, <A HREF="../../../org/jgap/ICompositeGene.html" title="interface in org.jgap">ICompositeGene</A>, <A HREF="../../../org/jgap/IPersistentRepresentation.html" title="interface in org.jgap">IPersistentRepresentation</A>, <A HREF="../../../org/jgap/supergenes/Supergene.html" title="interface in org.jgap.supergenes">Supergene</A>, <A HREF="../../../org/jgap/supergenes/SupergeneValidator.html" title="interface in org.jgap.supergenes">SupergeneValidator</A></DD>
</DL>
<DL>
<DT><B>Direct Known Subclasses:</B> <DD><A HREF="../../../examples/supergene/NickelsPenniesSupergene.html" title="class in examples.supergene">NickelsPenniesSupergene</A></DD>
</DL>
<HR>
<DL>
<DT><PRE>public abstract class <B>AbstractSupergene</B><DT>extends <A HREF="../../../org/jgap/BaseGene.html" title="class in org.jgap">BaseGene</A><DT>implements <A HREF="../../../org/jgap/supergenes/Supergene.html" title="interface in org.jgap.supergenes">Supergene</A>, <A HREF="../../../org/jgap/supergenes/SupergeneValidator.html" title="interface in org.jgap.supergenes">SupergeneValidator</A>, <A HREF="../../../org/jgap/IPersistentRepresentation.html" title="interface in org.jgap">IPersistentRepresentation</A></DL>
</PRE>

<P>
Combined implementation of both Supergene and SupergeneValidator.
 A working supergene can be easily created from this class just by
 adding genes and overriding
 <A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#isValid(org.jgap.Gene[], org.jgap.supergenes.Supergene)"><CODE>isValid (Gene [], Supergene)</CODE></A>
  method. For more complex cases, you may need to set your own
 <A HREF="../../../org/jgap/supergenes/Validator.html" title="class in org.jgap.supergenes"><CODE>Validator</CODE></A>.
<P>

<P>
<DL>
<DT><B>Since:</B></DT>
  <DD>2.0</DD>
<DT><B>See Also:</B><DD><A HREF="../../../serialized-form.html#org.jgap.supergenes.AbstractSupergene">Serialized Form</A></DL>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#GENE_DELIMITER">GENE_DELIMITER</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This field separates gene class name from
 the gene persistent representation string.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#GENE_DELIMITER_CLOSING">GENE_DELIMITER_CLOSING</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Represents the closing delimiter that is used to separate genes in the
 persistent representation of CompositeGene instances.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#GENE_DELIMITER_HEADING">GENE_DELIMITER_HEADING</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Represents the heading delimiter that is used to separate genes in the
 persistent representation of CompositeGene instances.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../org/jgap/supergenes/SupergeneValidator.html" title="interface in org.jgap.supergenes">SupergeneValidator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#m_validator">m_validator</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A validator (initially set to <i>this</i></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#MAX_IMMUTABLE_GENES">MAX_IMMUTABLE_GENES</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Maximal number of notes about immutable genes per
 single gene position</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#MAX_RETRIES">MAX_RETRIES</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Maximal number of retries for applyMutation and setToRandomValue.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="fields_inherited_from_class_org.jgap.BaseGene"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Fields inherited from class org.jgap.<A HREF="../../../org/jgap/BaseGene.html" title="class in org.jgap">BaseGene</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../org/jgap/BaseGene.html#DELTA">DELTA</A>, <A HREF="../../../org/jgap/BaseGene.html#S_APPLICATION_DATA">S_APPLICATION_DATA</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="fields_inherited_from_class_org.jgap.Gene"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Fields inherited from interface org.jgap.<A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../org/jgap/Gene.html#PERSISTENT_FIELD_DELIMITER">PERSISTENT_FIELD_DELIMITER</A></CODE></TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#AbstractSupergene()">AbstractSupergene</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Default constructor for dynamic instantiation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#AbstractSupergene(org.jgap.Configuration)">AbstractSupergene</A></B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_config)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructor for dynamic instantiation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#AbstractSupergene(org.jgap.Configuration, org.jgap.Gene[])">AbstractSupergene</A></B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_conf,
                  <A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>[]&nbsp;a_genes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs abstract supergene with the given gene list.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#addGene(org.jgap.Gene)">addGene</A></B>(<A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>&nbsp;a_gene)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Append a new gene to the gene array.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#applyMutation(int, double)">applyMutation</A></B>(int&nbsp;a_index,
              double&nbsp;a_percentage)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Applies a mutation of a given intensity (percentage) onto the gene
 at the given index.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#cleanup()">cleanup</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calls cleanup() for each subgene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#compareTo(java.lang.Object)">compareTo</A></B>(java.lang.Object&nbsp;o)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calls compareTo() for all subgenes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#createGene(java.lang.String, java.lang.String)">createGene</A></B>(java.lang.String&nbsp;a_geneClassName,
           java.lang.String&nbsp;a_persistentRepresentation)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new instance of gene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../org/jgap/supergenes/SupergeneValidator.html" title="interface in org.jgap.supergenes">SupergeneValidator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#createValidator(java.lang.String)">createValidator</A></B>(java.lang.String&nbsp;a_rep)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Create validator from the string representation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#equals(java.lang.Object)">equals</A></B>(java.lang.Object&nbsp;a_gene)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calls equals() for each pair of genes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#geneAt(int)">geneAt</A></B>(int&nbsp;a_index)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the Gene at the given index (locus) within the Chromosome.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#getAllele()">getAllele</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves the allele value represented by this Supergene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#getGenes()">getGenes</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Get the array of genes - components of this supergene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#getInternalValue()">getInternalValue</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Each Gene implementation holds its own m_value object keeping the allele
 value.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#getPersistent()">getPersistent</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 The default implementation returns an empty string.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#getPersistentRepresentation()">getPersistentRepresentation</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves a string representation of the value of this Gene instance
 that includes any information required to reconstruct it at a later
 time, such as its value and internal state.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../../org/jgap/supergenes/SupergeneValidator.html" title="interface in org.jgap.supergenes">SupergeneValidator</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#getValidator()">getValidator</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gets an object, responsible for deciding if the Supergene allele
 combination is valid.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#hashCode()">hashCode</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns sum of hashCode() of the genes-components.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#isValid()">isValid</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Test the allele combination of this supergene for validity.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#isValid(org.jgap.Gene[], org.jgap.supergenes.Supergene)">isValid</A></B>(<A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>[]&nbsp;a_case,
        <A HREF="../../../org/jgap/supergenes/Supergene.html" title="interface in org.jgap.supergenes">Supergene</A>&nbsp;a_forSupergene)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Test the given gene list for validity.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#newGeneInternal()">newGeneInternal</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new instance of this Supergene class with the same number of
 genes, calling newGene() for each subgene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#reset()">reset</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Discards all internal caches, ensuring correct repetetive tests
 of performance.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#setAllele(java.lang.Object)">setAllele</A></B>(java.lang.Object&nbsp;a_superAllele)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the allele.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#setFromPersistent(java.lang.String)">setFromPersistent</A></B>(java.lang.String&nbsp;a_from)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Set a persistend string representation (if needed) for this validator.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#setToRandomValue(org.jgap.RandomGenerator)">setToRandomValue</A></B>(<A HREF="../../../org/jgap/RandomGenerator.html" title="interface in org.jgap">RandomGenerator</A>&nbsp;a_numberGenerator)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the value of this Gene to a random legal value for the
 implementation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#setValidator(org.jgap.supergenes.SupergeneValidator)">setValidator</A></B>(<A HREF="../../../org/jgap/supergenes/SupergeneValidator.html" title="interface in org.jgap.supergenes">SupergeneValidator</A>&nbsp;a_validator)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets an object, responsible for deciding if the Supergene allele
 combination is valid.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#setValueFromPersistentRepresentation(java.lang.String)">setValueFromPersistentRepresentation</A></B>(java.lang.String&nbsp;a_representation)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the value and internal state of this Gene from the string
 representation returned by a previous invocation of the
 getPersistentRepresentation() method.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#size()">size</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the number of the genes-components of this supergene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected static&nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#split(java.lang.String)">split</A></B>(java.lang.String&nbsp;a_string)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Splits the string a_x into individual gene representations</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/supergenes/AbstractSupergene.html#toString()">toString</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves a string representation of this Gene's value that
 may be useful for display purposes.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_org.jgap.BaseGene"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class org.jgap.<A HREF="../../../org/jgap/BaseGene.html" title="class in org.jgap">BaseGene</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../org/jgap/BaseGene.html#compareApplicationData(java.lang.Object, java.lang.Object)">compareApplicationData</A>, <A HREF="../../../org/jgap/BaseGene.html#decode(java.lang.String)">decode</A>, <A HREF="../../../org/jgap/BaseGene.html#encode(java.lang.String)">encode</A>, <A HREF="../../../org/jgap/BaseGene.html#getApplicationData()">getApplicationData</A>, <A HREF="../../../org/jgap/BaseGene.html#getBusinessKey()">getBusinessKey</A>, <A HREF="../../../org/jgap/BaseGene.html#getConfiguration()">getConfiguration</A>, <A HREF="../../../org/jgap/BaseGene.html#getConstraintChecker()">getConstraintChecker</A>, <A HREF="../../../org/jgap/BaseGene.html#getEnergy()">getEnergy</A>, <A HREF="../../../org/jgap/BaseGene.html#isCompareApplicationData()">isCompareApplicationData</A>, <A HREF="../../../org/jgap/BaseGene.html#newGene()">newGene</A>, <A HREF="../../../org/jgap/BaseGene.html#setApplicationData(java.lang.Object)">setApplicationData</A>, <A HREF="../../../org/jgap/BaseGene.html#setCompareApplicationData(boolean)">setCompareApplicationData</A>, <A HREF="../../../org/jgap/BaseGene.html#setConstraintChecker(org.jgap.IGeneConstraintChecker)">setConstraintChecker</A>, <A HREF="../../../org/jgap/BaseGene.html#setEnergy(double)">setEnergy</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, finalize, getClass, notify, notifyAll, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_org.jgap.Gene"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from interface org.jgap.<A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../../org/jgap/Gene.html#getApplicationData()">getApplicationData</A>, <A HREF="../../../org/jgap/Gene.html#getConfiguration()">getConfiguration</A>, <A HREF="../../../org/jgap/Gene.html#getEnergy()">getEnergy</A>, <A HREF="../../../org/jgap/Gene.html#isCompareApplicationData()">isCompareApplicationData</A>, <A HREF="../../../org/jgap/Gene.html#newGene()">newGene</A>, <A HREF="../../../org/jgap/Gene.html#setApplicationData(java.lang.Object)">setApplicationData</A>, <A HREF="../../../org/jgap/Gene.html#setCompareApplicationData(boolean)">setCompareApplicationData</A>, <A HREF="../../../org/jgap/Gene.html#setConstraintChecker(org.jgap.IGeneConstraintChecker)">setConstraintChecker</A>, <A HREF="../../../org/jgap/Gene.html#setEnergy(double)">setEnergy</A></CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="GENE_DELIMITER"><!-- --></A><H3>
GENE_DELIMITER</H3>
<PRE>
public static final java.lang.String <B>GENE_DELIMITER</B></PRE>
<DL>
<DD>This field separates gene class name from
 the gene persistent representation string.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../constant-values.html#org.jgap.supergenes.AbstractSupergene.GENE_DELIMITER">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="GENE_DELIMITER_HEADING"><!-- --></A><H3>
GENE_DELIMITER_HEADING</H3>
<PRE>
public static final java.lang.String <B>GENE_DELIMITER_HEADING</B></PRE>
<DL>
<DD>Represents the heading delimiter that is used to separate genes in the
 persistent representation of CompositeGene instances.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../constant-values.html#org.jgap.supergenes.AbstractSupergene.GENE_DELIMITER_HEADING">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="GENE_DELIMITER_CLOSING"><!-- --></A><H3>
GENE_DELIMITER_CLOSING</H3>
<PRE>
public static final java.lang.String <B>GENE_DELIMITER_CLOSING</B></PRE>
<DL>
<DD>Represents the closing delimiter that is used to separate genes in the
 persistent representation of CompositeGene instances.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../constant-values.html#org.jgap.supergenes.AbstractSupergene.GENE_DELIMITER_CLOSING">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="MAX_RETRIES"><!-- --></A><H3>
MAX_RETRIES</H3>
<PRE>
public static final int <B>MAX_RETRIES</B></PRE>
<DL>
<DD>Maximal number of retries for applyMutation and setToRandomValue.
 If the valid supergen cannot be created after this number of iterations,
 the error message is printed and the unchanged instance is returned.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../constant-values.html#org.jgap.supergenes.AbstractSupergene.MAX_RETRIES">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="MAX_IMMUTABLE_GENES"><!-- --></A><H3>
MAX_IMMUTABLE_GENES</H3>
<PRE>
public static final int <B>MAX_IMMUTABLE_GENES</B></PRE>
<DL>
<DD>Maximal number of notes about immutable genes per
 single gene position
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../../constant-values.html#org.jgap.supergenes.AbstractSupergene.MAX_IMMUTABLE_GENES">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="m_validator"><!-- --></A><H3>
m_validator</H3>
<PRE>
protected <A HREF="../../../org/jgap/supergenes/SupergeneValidator.html" title="interface in org.jgap.supergenes">SupergeneValidator</A> <B>m_validator</B></PRE>
<DL>
<DD>A validator (initially set to <i>this</i>
<P>
<DL>
</DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="AbstractSupergene()"><!-- --></A><H3>
AbstractSupergene</H3>
<PRE>
public <B>AbstractSupergene</B>()
                  throws <A HREF="../../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DD>Default constructor for dynamic instantiation.
<P>
<DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE><DT><B>Since:</B></DT>
  <DD>3.0</DD>
</DL>
</DL>
<HR>

<A NAME="AbstractSupergene(org.jgap.Configuration)"><!-- --></A><H3>
AbstractSupergene</H3>
<PRE>
public <B>AbstractSupergene</B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_config)
                  throws <A HREF="../../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DD>Constructor for dynamic instantiation.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>a_config</CODE> - the configuration to use
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE><DT><B>Since:</B></DT>
  <DD>3.0</DD>
</DL>
</DL>
<HR>

<A NAME="AbstractSupergene(org.jgap.Configuration, org.jgap.Gene[])"><!-- --></A><H3>
AbstractSupergene</H3>
<PRE>
public <B>AbstractSupergene</B>(<A HREF="../../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_conf,
                         <A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>[]&nbsp;a_genes)
                  throws <A HREF="../../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DD>Constructs abstract supergene with the given gene list.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>a_conf</CODE> - the configuration to use<DD><CODE>a_genes</CODE> - array of genes for this Supergene
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE></DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="getGenes()"><!-- --></A><H3>
getGenes</H3>
<PRE>
public <A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>[] <B>getGenes</B>()</PRE>
<DL>
<DD><B>Description copied from interface: <CODE><A HREF="../../../org/jgap/supergenes/Supergene.html#getGenes()">Supergene</A></CODE></B></DD>
<DD>Get the array of genes - components of this supergene.
 The supergene components may be supergenes itself.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/supergenes/Supergene.html#getGenes()">getGenes</A></CODE> in interface <CODE><A HREF="../../../org/jgap/supergenes/Supergene.html" title="interface in org.jgap.supergenes">Supergene</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the array of genes - components of this supergene. The supergene
 components may be supergenes itself</DL>
</DD>
</DL>
<HR>

<A NAME="geneAt(int)"><!-- --></A><H3>
geneAt</H3>
<PRE>
public final <A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A> <B>geneAt</B>(int&nbsp;a_index)</PRE>
<DL>
<DD>Returns the Gene at the given index (locus) within the Chromosome. The
 first gene is at index zero and the last gene is at the index equal to
 the size of this Chromosome - 1.

 This seems to be one of the bottlenecks, so it is declared final.
 I cannot imagine the reason for overriding this trivial single line
 method.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/ICompositeGene.html#geneAt(int)">geneAt</A></CODE> in interface <CODE><A HREF="../../../org/jgap/ICompositeGene.html" title="interface in org.jgap">ICompositeGene</A></CODE><DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/supergenes/Supergene.html#geneAt(int)">geneAt</A></CODE> in interface <CODE><A HREF="../../../org/jgap/supergenes/Supergene.html" title="interface in org.jgap.supergenes">Supergene</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_index</CODE> - the index of the gene value to be returned
<DT><B>Returns:</B><DD>the Gene at the given index</DL>
</DD>
</DL>
<HR>

<A NAME="isValid()"><!-- --></A><H3>
isValid</H3>
<PRE>
public boolean <B>isValid</B>()</PRE>
<DL>
<DD>Test the allele combination of this supergene for validity. This method
 calls isValid for the current gene list.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/supergenes/Supergene.html#isValid()">isValid</A></CODE> in interface <CODE><A HREF="../../../org/jgap/supergenes/Supergene.html" title="interface in org.jgap.supergenes">Supergene</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>true only if the supergene allele combination is valid
 or the setValidator (<i>null</i>) has been previously called</DL>
</DD>
</DL>
<HR>

<A NAME="isValid(org.jgap.Gene[], org.jgap.supergenes.Supergene)"><!-- --></A><H3>
isValid</H3>
<PRE>
public boolean <B>isValid</B>(<A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>[]&nbsp;a_case,
                       <A HREF="../../../org/jgap/supergenes/Supergene.html" title="interface in org.jgap.supergenes">Supergene</A>&nbsp;a_forSupergene)</PRE>
<DL>
<DD>Test the given gene list for validity. The genes must exactly the same
 as inside this supergene.
 At <i>least about 5 % of the randomly
 generated Supergene suparallele values should be valid.</i> If the valid
 combinations represents too small part of all possible combinations,
 it can take too long to find the suitable mutation that does not brake
 a supergene. If you face this problem, try to split the supergene into
 several sub-supergenes.

 This method is only called if you have not set any alternative
 validator (including <i>null</i>).
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/supergenes/SupergeneValidator.html#isValid(org.jgap.Gene[], org.jgap.supergenes.Supergene)">isValid</A></CODE> in interface <CODE><A HREF="../../../org/jgap/supergenes/SupergeneValidator.html" title="interface in org.jgap.supergenes">SupergeneValidator</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_case</CODE> - ignored here<DD><CODE>a_forSupergene</CODE> - ignored here
<DT><B>Returns:</B><DD>true only if the supergene allele combination is valid
<DT><B>Throws:</B>
<DD><CODE>java.lang.Error</CODE> - by default. If you do not set external validator,
 you should always override this method</DL>
</DD>
</DL>
<HR>

<A NAME="newGeneInternal()"><!-- --></A><H3>
newGeneInternal</H3>
<PRE>
protected <A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A> <B>newGeneInternal</B>()</PRE>
<DL>
<DD>Creates a new instance of this Supergene class with the same number of
 genes, calling newGene() for each subgene. The class, derived from this
 abstract supergene will be instantiated
 (not the instance of abstractSupergene itself). If the external
 validator is set, the same validator will be set for the new gene.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/BaseGene.html#newGeneInternal()">newGeneInternal</A></CODE> in class <CODE><A HREF="../../../org/jgap/BaseGene.html" title="class in org.jgap">BaseGene</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the new Gene
<DT><B>Throws:</B>
<DD><CODE>java.lang.Error</CODE> - if the instance of <i>this</i> cannot be instantiated
 (for example, if it is not public or  the parameterless constructor is
 not provided).</DL>
</DD>
</DL>
<HR>

<A NAME="applyMutation(int, double)"><!-- --></A><H3>
applyMutation</H3>
<PRE>
public void <B>applyMutation</B>(int&nbsp;a_index,
                          double&nbsp;a_percentage)</PRE>
<DL>
<DD>Applies a mutation of a given intensity (percentage) onto the gene
 at the given index. Retries while isValid() returns true for the
 supergene. The method is delegated to the first element ] of the
 gene, indexed by <code>index</code>.
 See org.jgap.supergenes.AbstractSupergene.isValid()
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/Gene.html#applyMutation(int, double)">applyMutation</A></CODE> in interface <CODE><A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_index</CODE> - index of atomic element, between 0 and size()-1<DD><CODE>a_percentage</CODE> - percentage of mutation (greater than -1 and smaller
 than 1).</DL>
</DD>
</DL>
<HR>

<A NAME="reset()"><!-- --></A><H3>
reset</H3>
<PRE>
public static void <B>reset</B>()</PRE>
<DL>
<DD>Discards all internal caches, ensuring correct repetetive tests
 of performance. Differently from cleanup(), discards also static
 references, that are assumed to be useful for the multiple instances
 of the Supergene.
 Clears the set of the alleles that are known to be immutable.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setToRandomValue(org.jgap.RandomGenerator)"><!-- --></A><H3>
setToRandomValue</H3>
<PRE>
public void <B>setToRandomValue</B>(<A HREF="../../../org/jgap/RandomGenerator.html" title="interface in org.jgap">RandomGenerator</A>&nbsp;a_numberGenerator)</PRE>
<DL>
<DD>Sets the value of this Gene to a random legal value for the
 implementation. It calls setToRandomValue for all subgenes and
 then validates. With a large number of subgenes and low percent of
 valid combinations this may take too long to complete. We think,
 at lease several % of the all possible combintations must be valid.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/Gene.html#setToRandomValue(org.jgap.RandomGenerator)">setToRandomValue</A></CODE> in interface <CODE><A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_numberGenerator</CODE> - The random number generator that should be used
 to create any random values. It's important to use this generator to
 maintain the user's flexibility to configure the genetic engine to use the
 random number generator of their choice</DL>
</DD>
</DL>
<HR>

<A NAME="setAllele(java.lang.Object)"><!-- --></A><H3>
setAllele</H3>
<PRE>
public void <B>setAllele</B>(java.lang.Object&nbsp;a_superAllele)</PRE>
<DL>
<DD>Sets the allele.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/Gene.html#setAllele(java.lang.Object)">setAllele</A></CODE> in interface <CODE><A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_superAllele</CODE> - must be an array of objects, size matching the
 number of genes</DL>
</DD>
</DL>
<HR>

<A NAME="getAllele()"><!-- --></A><H3>
getAllele</H3>
<PRE>
public java.lang.Object <B>getAllele</B>()</PRE>
<DL>
<DD>Retrieves the allele value represented by this Supergene.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/Gene.html#getAllele()">getAllele</A></CODE> in interface <CODE><A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../../org/jgap/BaseGene.html#getAllele()">getAllele</A></CODE> in class <CODE><A HREF="../../../org/jgap/BaseGene.html" title="class in org.jgap">BaseGene</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>array of objects, each matching the subgene in this Supergene</DL>
</DD>
</DL>
<HR>

<A NAME="getPersistentRepresentation()"><!-- --></A><H3>
getPersistentRepresentation</H3>
<PRE>
public java.lang.String <B>getPersistentRepresentation</B>()
                                             throws java.lang.UnsupportedOperationException</PRE>
<DL>
<DD><B>Description copied from interface: <CODE><A HREF="../../../org/jgap/Gene.html#getPersistentRepresentation()">Gene</A></CODE></B></DD>
<DD>Retrieves a string representation of the value of this Gene instance
 that includes any information required to reconstruct it at a later
 time, such as its value and internal state. This string will be used to
 represent this Gene instance in XML persistence. This is an optional
 method but, if not implemented, XML persistence and possibly other
 features will not be available. An UnsupportedOperationException should
 be thrown if no implementation is provided.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/Gene.html#getPersistentRepresentation()">getPersistentRepresentation</A></CODE> in interface <CODE><A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A></CODE><DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/IPersistentRepresentation.html#getPersistentRepresentation()">getPersistentRepresentation</A></CODE> in interface <CODE><A HREF="../../../org/jgap/IPersistentRepresentation.html" title="interface in org.jgap">IPersistentRepresentation</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>a string representation of the value of this Supergene
 instance, using calls to the Supergene components. Supports other
 (nested) supergenes in this supergene
<DT><B>Throws:</B>
<DD><CODE>java.lang.UnsupportedOperationException</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="setValueFromPersistentRepresentation(java.lang.String)"><!-- --></A><H3>
setValueFromPersistentRepresentation</H3>
<PRE>
public void <B>setValueFromPersistentRepresentation</B>(java.lang.String&nbsp;a_representation)
                                          throws <A HREF="../../../org/jgap/UnsupportedRepresentationException.html" title="class in org.jgap">UnsupportedRepresentationException</A></PRE>
<DL>
<DD>Sets the value and internal state of this Gene from the string
 representation returned by a previous invocation of the
 getPersistentRepresentation() method.

 If the validator is not THIS and not null, a new validator is
 created using Class.forName(..).newInstance.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/Gene.html#setValueFromPersistentRepresentation(java.lang.String)">setValueFromPersistentRepresentation</A></CODE> in interface <CODE><A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A></CODE><DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/IPersistentRepresentation.html#setValueFromPersistentRepresentation(java.lang.String)">setValueFromPersistentRepresentation</A></CODE> in interface <CODE><A HREF="../../../org/jgap/IPersistentRepresentation.html" title="interface in org.jgap">IPersistentRepresentation</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_representation</CODE> - the string representation retrieved from a
 prior call to the getPersistentRepresentation() method
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../org/jgap/UnsupportedRepresentationException.html" title="class in org.jgap">UnsupportedRepresentationException</A></CODE><DT><B>Since:</B></DT>
  <DD>2.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="createValidator(java.lang.String)"><!-- --></A><H3>
createValidator</H3>
<PRE>
protected <A HREF="../../../org/jgap/supergenes/SupergeneValidator.html" title="interface in org.jgap.supergenes">SupergeneValidator</A> <B>createValidator</B>(java.lang.String&nbsp;a_rep)</PRE>
<DL>
<DD>Create validator from the string representation.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="createGene(java.lang.String, java.lang.String)"><!-- --></A><H3>
createGene</H3>
<PRE>
protected <A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A> <B>createGene</B>(java.lang.String&nbsp;a_geneClassName,
                          java.lang.String&nbsp;a_persistentRepresentation)
                   throws java.lang.Exception</PRE>
<DL>
<DD>Creates a new instance of gene.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE>java.lang.Exception</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="cleanup()"><!-- --></A><H3>
cleanup</H3>
<PRE>
public void <B>cleanup</B>()</PRE>
<DL>
<DD>Calls cleanup() for each subgene.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/Gene.html#cleanup()">cleanup</A></CODE> in interface <CODE><A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../../org/jgap/BaseGene.html#cleanup()">cleanup</A></CODE> in class <CODE><A HREF="../../../org/jgap/BaseGene.html" title="class in org.jgap">BaseGene</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="toString()"><!-- --></A><H3>
toString</H3>
<PRE>
public java.lang.String <B>toString</B>()</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../../org/jgap/BaseGene.html#toString()">BaseGene</A></CODE></B></DD>
<DD>Retrieves a string representation of this Gene's value that
 may be useful for display purposes.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/Gene.html#toString()">toString</A></CODE> in interface <CODE><A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../../org/jgap/BaseGene.html#toString()">toString</A></CODE> in class <CODE><A HREF="../../../org/jgap/BaseGene.html" title="class in org.jgap">BaseGene</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>a string representation of the supergene, providing
 class name and calling toString() for all subgenes.</DL>
</DD>
</DL>
<HR>

<A NAME="size()"><!-- --></A><H3>
size</H3>
<PRE>
public int <B>size</B>()</PRE>
<DL>
<DD>Returns the number of the genes-components of this supergene.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/Gene.html#size()">size</A></CODE> in interface <CODE><A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../../org/jgap/BaseGene.html#size()">size</A></CODE> in class <CODE><A HREF="../../../org/jgap/BaseGene.html" title="class in org.jgap">BaseGene</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the size of the gene, i.e the number of atomic elements. Always 1
 for non-composed Gene types. Override for composed Gene types</DL>
</DD>
</DL>
<HR>

<A NAME="compareTo(java.lang.Object)"><!-- --></A><H3>
compareTo</H3>
<PRE>
public int <B>compareTo</B>(java.lang.Object&nbsp;o)</PRE>
<DL>
<DD>Calls compareTo() for all subgenes. The passed parameter must be
 an instance of AbstractSupergene.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE>compareTo</CODE> in interface <CODE>java.lang.Comparable</CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="equals(java.lang.Object)"><!-- --></A><H3>
equals</H3>
<PRE>
public boolean <B>equals</B>(java.lang.Object&nbsp;a_gene)</PRE>
<DL>
<DD>Calls equals() for each pair of genes. If the supplied object is
 an instance of the different class, returns false. Also, the
 genes are assumed to be different if they have different validator
 classes (or only one of the validators is set to null).
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../../org/jgap/BaseGene.html#equals(java.lang.Object)">equals</A></CODE> in class <CODE><A HREF="../../../org/jgap/BaseGene.html" title="class in org.jgap">BaseGene</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_gene</CODE> - the object to compare to this Gene for equality
<DT><B>Returns:</B><DD>true if this Gene is equal to the given object, false otherwise</DL>
</DD>
</DL>
<HR>

<A NAME="hashCode()"><!-- --></A><H3>
hashCode</H3>
<PRE>
public int <B>hashCode</B>()</PRE>
<DL>
<DD>Returns sum of hashCode() of the genes-components.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE><A HREF="../../../org/jgap/BaseGene.html#hashCode()">hashCode</A></CODE> in class <CODE><A HREF="../../../org/jgap/BaseGene.html" title="class in org.jgap">BaseGene</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>this Gene's hash code</DL>
</DD>
</DL>
<HR>

<A NAME="split(java.lang.String)"><!-- --></A><H3>
split</H3>
<PRE>
protected static final java.util.List <B>split</B>(java.lang.String&nbsp;a_string)
                                     throws <A HREF="../../../org/jgap/UnsupportedRepresentationException.html" title="class in org.jgap">UnsupportedRepresentationException</A></PRE>
<DL>
<DD>Splits the string a_x into individual gene representations
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_string</CODE> - the string to split
<DT><B>Returns:</B><DD>the elements of the returned array are the
 persistent representation strings of the genes - components
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../../org/jgap/UnsupportedRepresentationException.html" title="class in org.jgap">UnsupportedRepresentationException</A></CODE></DL>
</DD>
</DL>
<HR>

<A NAME="addGene(org.jgap.Gene)"><!-- --></A><H3>
addGene</H3>
<PRE>
public void <B>addGene</B>(<A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>&nbsp;a_gene)</PRE>
<DL>
<DD>Append a new gene to the gene array.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/ICompositeGene.html#addGene(org.jgap.Gene)">addGene</A></CODE> in interface <CODE><A HREF="../../../org/jgap/ICompositeGene.html" title="interface in org.jgap">ICompositeGene</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_gene</CODE> - the gene to add</DL>
</DD>
</DL>
<HR>

<A NAME="setValidator(org.jgap.supergenes.SupergeneValidator)"><!-- --></A><H3>
setValidator</H3>
<PRE>
public void <B>setValidator</B>(<A HREF="../../../org/jgap/supergenes/SupergeneValidator.html" title="interface in org.jgap.supergenes">SupergeneValidator</A>&nbsp;a_validator)</PRE>
<DL>
<DD>Sets an object, responsible for deciding if the Supergene allele
 combination is valid. If it is set to null, no validation is performed
 (all combinations are assumed to be valid). If no validator is
 set, the method <code>isValid (Gene [] ) </code>is called.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/supergenes/Supergene.html#setValidator(org.jgap.supergenes.SupergeneValidator)">setValidator</A></CODE> in interface <CODE><A HREF="../../../org/jgap/supergenes/Supergene.html" title="interface in org.jgap.supergenes">Supergene</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getValidator()"><!-- --></A><H3>
getValidator</H3>
<PRE>
public <A HREF="../../../org/jgap/supergenes/SupergeneValidator.html" title="interface in org.jgap.supergenes">SupergeneValidator</A> <B>getValidator</B>()</PRE>
<DL>
<DD>Gets an object, responsible for deciding if the Supergene allele
 combination is valid. If no external validator was set and the
 class uses its own internal validation method, it returns <i>this</i>
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/supergenes/Supergene.html#getValidator()">getValidator</A></CODE> in interface <CODE><A HREF="../../../org/jgap/supergenes/Supergene.html" title="interface in org.jgap.supergenes">Supergene</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getPersistent()"><!-- --></A><H3>
getPersistent</H3>
<PRE>
public java.lang.String <B>getPersistent</B>()</PRE>
<DL>
<DD>
 The default implementation returns an empty string.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/supergenes/SupergeneValidator.html#getPersistent()">getPersistent</A></CODE> in interface <CODE><A HREF="../../../org/jgap/supergenes/SupergeneValidator.html" title="interface in org.jgap.supergenes">SupergeneValidator</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>persistent string representation (if needed) of this validator.
 The method name is different allowing the same class to implement both
 Supergene and supergeneValidator.</DL>
</DD>
</DL>
<HR>

<A NAME="setFromPersistent(java.lang.String)"><!-- --></A><H3>
setFromPersistent</H3>
<PRE>
public void <B>setFromPersistent</B>(java.lang.String&nbsp;a_from)</PRE>
<DL>
<DD>Set a persistend string representation (if needed) for this validator.
 The method name is different allowing the same class to implement both
 Supergene and supergeneValidator.
 The default implementation does nothing.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/supergenes/SupergeneValidator.html#setFromPersistent(java.lang.String)">setFromPersistent</A></CODE> in interface <CODE><A HREF="../../../org/jgap/supergenes/SupergeneValidator.html" title="interface in org.jgap.supergenes">SupergeneValidator</A></CODE></DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getInternalValue()"><!-- --></A><H3>
getInternalValue</H3>
<PRE>
public java.lang.Object <B>getInternalValue</B>()</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../../org/jgap/BaseGene.html#getInternalValue()">BaseGene</A></CODE></B></DD>
<DD>Each Gene implementation holds its own m_value object keeping the allele
 value. In your Gene implementation, just return it with this method
 (see <A HREF="../../../org/jgap/impl/BooleanGene.html" title="class in org.jgap.impl"><CODE>BooleanGene</CODE></A> for example)
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/BaseGene.html#getInternalValue()">getInternalValue</A></CODE> in class <CODE><A HREF="../../../org/jgap/BaseGene.html" title="class in org.jgap">BaseGene</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>not needed for abstract supergene</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
JGAP</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../../org/jgap/supergenes/Supergene.html" title="interface in org.jgap.supergenes"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html?org/jgap/supergenes/AbstractSupergene.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="AbstractSupergene.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
