<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_02) on Mon Nov 05 17:25:02 CET 2007 -->
<TITLE>
Chromosome
</TITLE>

<META NAME="date" CONTENT="2007-11-05">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="Chromosome";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
JGAP</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../org/jgap/CachedFitnessFunction.html" title="class in org.jgap"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?org/jgap/Chromosome.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Chromosome.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.jgap</FONT>
<BR>
Class Chromosome</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../resources/inherit.gif" ALT="extended by "><A HREF="../../org/jgap/BaseChromosome.html" title="class in org.jgap">org.jgap.BaseChromosome</A>
      <IMG SRC="../../resources/inherit.gif" ALT="extended by "><B>org.jgap.Chromosome</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD>java.io.Serializable, java.lang.Cloneable, java.lang.Comparable, <A HREF="../../org/jgap/IBusinessKey.html" title="interface in org.jgap">IBusinessKey</A>, <A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A>, <A HREF="../../org/jgap/IHandler.html" title="interface in org.jgap">IHandler</A>, <A HREF="../../org/jgap/IInitializer.html" title="interface in org.jgap">IInitializer</A>, <A HREF="../../org/jgap/IPersistentRepresentation.html" title="interface in org.jgap">IPersistentRepresentation</A>, <A HREF="../../org/jgap/util/ICloneable.html" title="interface in org.jgap.util">ICloneable</A></DD>
</DL>
<DL>
<DT><B>Direct Known Subclasses:</B> <DD><A HREF="../../examples/distinctGenes/MyChromosome.html" title="class in examples.distinctGenes">MyChromosome</A></DD>
</DL>
<HR>
<DL>
<DT><PRE>public class <B>Chromosome</B><DT>extends <A HREF="../../org/jgap/BaseChromosome.html" title="class in org.jgap">BaseChromosome</A></DL>
</PRE>

<P>
Chromosomes represent potential solutions and consist of a fixed-length
 collection of genes. Each gene represents a discrete part of the solution.
 Each gene in the Chromosome may be backed by a different concrete
 implementation of the Gene interface, but all genes in a respective
 position (locus) must share the same concrete implementation across
 Chromosomes within a single population (genotype). In other words, gene 1
 in a chromosome must share the same concrete implementation as gene 1 in all
 other chromosomes in the population.
<P>

<P>
<DL>
<DT><B>Since:</B></DT>
  <DD>1.0</DD>
<DT><B>See Also:</B><DD><A HREF="../../serialized-form.html#org.jgap.Chromosome">Serialized Form</A></DL>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#m_fitnessValue">m_fitnessValue</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Stores the fitness value of this Chromosome as determined by the
 active fitness function.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="fields_inherited_from_class_org.jgap.BaseChromosome"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Fields inherited from class org.jgap.<A HREF="../../org/jgap/BaseChromosome.html" title="class in org.jgap">BaseChromosome</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../org/jgap/BaseChromosome.html#CHROM_DELIMITER">CHROM_DELIMITER</A>, <A HREF="../../org/jgap/BaseChromosome.html#GENE_DELIMITER">GENE_DELIMITER</A>, <A HREF="../../org/jgap/BaseChromosome.html#GENE_DELIMITER_CLOSING">GENE_DELIMITER_CLOSING</A>, <A HREF="../../org/jgap/BaseChromosome.html#GENE_DELIMITER_HEADING">GENE_DELIMITER_HEADING</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="fields_inherited_from_class_org.jgap.IChromosome"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Fields inherited from interface org.jgap.<A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../org/jgap/IChromosome.html#S_ALLELES">S_ALLELES</A>, <A HREF="../../org/jgap/IChromosome.html#S_APPLICATION_DATA">S_APPLICATION_DATA</A>, <A HREF="../../org/jgap/IChromosome.html#S_FITNESS_VALUE">S_FITNESS_VALUE</A>, <A HREF="../../org/jgap/IChromosome.html#S_SIZE">S_SIZE</A></CODE></TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#Chromosome()">Chromosome</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Default constructor, only provided for dynamic instantiation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#Chromosome(org.jgap.Configuration)">Chromosome</A></B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_configuration)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructor, provided for dynamic or minimal instantiation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#Chromosome(org.jgap.Configuration, org.jgap.Gene[])">Chromosome</A></B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_configuration,
           <A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>[]&nbsp;a_initialGenes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs a Chromosome separate from any specific Configuration.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#Chromosome(org.jgap.Configuration, org.jgap.Gene[], org.jgap.IGeneConstraintChecker)">Chromosome</A></B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_configuration,
           <A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>[]&nbsp;a_initialGenes,
           <A HREF="../../org/jgap/IGeneConstraintChecker.html" title="interface in org.jgap">IGeneConstraintChecker</A>&nbsp;a_constraintChecker)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs a Chromosome separate from any specific Configuration.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#Chromosome(org.jgap.Configuration, org.jgap.Gene, int)">Chromosome</A></B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_configuration,
           <A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>&nbsp;a_sampleGene,
           int&nbsp;a_desiredSize)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructs a Chromosome of the given size separate from any specific
 Configuration.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#Chromosome(org.jgap.Configuration, org.jgap.Gene, int, org.jgap.IGeneConstraintChecker)">Chromosome</A></B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_configuration,
           <A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>&nbsp;a_sampleGene,
           int&nbsp;a_desiredSize,
           <A HREF="../../org/jgap/IGeneConstraintChecker.html" title="interface in org.jgap">IGeneConstraintChecker</A>&nbsp;a_constraintChecker)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#Chromosome(org.jgap.Configuration, int)">Chromosome</A></B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_configuration,
           int&nbsp;a_desiredSize)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructor for specifying the number of genes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#Chromosome(org.jgap.Configuration, java.lang.String)">Chromosome</A></B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_configuration,
           java.lang.String&nbsp;a_persistentRepresentatuion)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructor, provided for instantiation via persistent representation.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#calcFitnessValue()">calcFitnessValue</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#checkGenes(org.jgap.Gene[])">checkGenes</A></B>(<A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>[]&nbsp;a_initialGenes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Helper: called by constructors only to verify the initial genes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#cleanup()">cleanup</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Invoked when this Chromosome is no longer needed and should perform
 any necessary cleanup.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#clone()">clone</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a copy of this Chromosome.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#cloneObject(java.lang.Object)">cloneObject</A></B>(java.lang.Object&nbsp;a_object)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Clones an object by using clone handlers.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#compareTo(java.lang.Object)">compareTo</A></B>(java.lang.Object&nbsp;other)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Compares the given Chromosome to this Chromosome.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#equals(java.lang.Object)">equals</A></B>(java.lang.Object&nbsp;other)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Compares this Chromosome against the specified object.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#getApplicationData()">getApplicationData</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves the application-specific data that is attached to this
 Chromosome.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../org/jgap/IGeneConstraintChecker.html" title="interface in org.jgap">IGeneConstraintChecker</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#getConstraintChecker()">getConstraintChecker</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#getFitnessValue()">getFitnessValue</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves the fitness value of this Chromosome, as determined by the
 active fitness function.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#getFitnessValueDirectly()">getFitnessValueDirectly</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#getMultiObjectives()">getMultiObjectives</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#hashCode()">hashCode</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieve a hash code for this Chromosome.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#initFromGene(org.jgap.Gene)">initFromGene</A></B>(<A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>&nbsp;a_sampleGene)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#isCompareApplicationData()">isCompareApplicationData</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#isHandlerFor(java.lang.Object, java.lang.Class)">isHandlerFor</A></B>(java.lang.Object&nbsp;a_obj,
             java.lang.Class&nbsp;a_class)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Determines whether the handler is suitable for the given object instance
 or class.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#isSelectedForNextGeneration()">isSelectedForNextGeneration</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves whether this Chromosome has been selected by the natural
 selector to continue to the next generation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#perform(java.lang.Object, java.lang.Class, java.lang.Object)">perform</A></B>(java.lang.Object&nbsp;a_obj,
        java.lang.Class&nbsp;a_class,
        java.lang.Object&nbsp;a_params)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Performs a task for the given object or class.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;<A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#randomInitialChromosome(org.jgap.Configuration)">randomInitialChromosome</A></B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_configuration)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Convenience method that returns a new Chromosome instance with its
 genes values (alleles) randomized.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#setApplicationData(java.lang.Object)">setApplicationData</A></B>(java.lang.Object&nbsp;a_newData)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This sets the application-specific data that is attached to this
 Chromosome.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#setCompareApplicationData(boolean)">setCompareApplicationData</A></B>(boolean&nbsp;a_doCompare)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Should we also consider the application data when comparing?</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#setConstraintChecker(org.jgap.IGeneConstraintChecker)">setConstraintChecker</A></B>(<A HREF="../../org/jgap/IGeneConstraintChecker.html" title="interface in org.jgap">IGeneConstraintChecker</A>&nbsp;a_constraintChecker)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the constraint checker to be used for this gene whenever method
 setAllele(Object) is called.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#setFitnessValue(double)">setFitnessValue</A></B>(double&nbsp;a_newFitnessValue)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the fitness value of this Chromosome.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#setFitnessValueDirectly(double)">setFitnessValueDirectly</A></B>(double&nbsp;a_newFitnessValue)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the fitness value of this Chromosome directly without any
 constraint checks, conversions or checks.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#setGenes(org.jgap.Gene[])">setGenes</A></B>(<A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>[]&nbsp;a_genes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the genes for the chromosome.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#setIsSelectedForNextGeneration(boolean)">setIsSelectedForNextGeneration</A></B>(boolean&nbsp;a_isSelected)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets whether this Chromosome has been selected by the natural selector
 to continue to the next generation or manually (e.g. via an add-method).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#setMultiObjectives(java.util.List)">setMultiObjectives</A></B>(java.util.List&nbsp;a_values)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#toString()">toString</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Chromosome.html#verify(org.jgap.IGeneConstraintChecker)">verify</A></B>(<A HREF="../../org/jgap/IGeneConstraintChecker.html" title="interface in org.jgap">IGeneConstraintChecker</A>&nbsp;a_constraintChecker)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Verifies the state of the chromosome.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_org.jgap.BaseChromosome"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class org.jgap.<A HREF="../../org/jgap/BaseChromosome.html" title="class in org.jgap">BaseChromosome</A></B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><A HREF="../../org/jgap/BaseChromosome.html#createGene(java.lang.String, java.lang.String)">createGene</A>, <A HREF="../../org/jgap/BaseChromosome.html#decode(java.lang.String)">decode</A>, <A HREF="../../org/jgap/BaseChromosome.html#encode(java.lang.String)">encode</A>, <A HREF="../../org/jgap/BaseChromosome.html#getAge()">getAge</A>, <A HREF="../../org/jgap/BaseChromosome.html#getBusinessKey()">getBusinessKey</A>, <A HREF="../../org/jgap/BaseChromosome.html#getConfiguration()">getConfiguration</A>, <A HREF="../../org/jgap/BaseChromosome.html#getGene(int)">getGene</A>, <A HREF="../../org/jgap/BaseChromosome.html#getGenes()">getGenes</A>, <A HREF="../../org/jgap/BaseChromosome.html#getGenesPersistentRepresentation()">getGenesPersistentRepresentation</A>, <A HREF="../../org/jgap/BaseChromosome.html#getGenesPersistentRepresentation(java.lang.StringBuffer)">getGenesPersistentRepresentation</A>, <A HREF="../../org/jgap/BaseChromosome.html#getPersistentRepresentation()">getPersistentRepresentation</A>, <A HREF="../../org/jgap/BaseChromosome.html#increaseAge()">increaseAge</A>, <A HREF="../../org/jgap/BaseChromosome.html#increaseOperatedOn()">increaseOperatedOn</A>, <A HREF="../../org/jgap/BaseChromosome.html#operatedOn()">operatedOn</A>, <A HREF="../../org/jgap/BaseChromosome.html#resetAge()">resetAge</A>, <A HREF="../../org/jgap/BaseChromosome.html#resetOperatedOn()">resetOperatedOn</A>, <A HREF="../../org/jgap/BaseChromosome.html#setGene(int, org.jgap.Gene)">setGene</A>, <A HREF="../../org/jgap/BaseChromosome.html#setValueFromPersistentRepresentation(java.lang.String)">setValueFromPersistentRepresentation</A>, <A HREF="../../org/jgap/BaseChromosome.html#size()">size</A>, <A HREF="../../org/jgap/BaseChromosome.html#split(java.lang.String)">split</A></CODE></TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>finalize, getClass, notify, notifyAll, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="m_fitnessValue"><!-- --></A><H3>
m_fitnessValue</H3>
<PRE>
protected double <B>m_fitnessValue</B></PRE>
<DL>
<DD>Stores the fitness value of this Chromosome as determined by the
 active fitness function. A value of -1 indicates that this field
 has not yet been set with this Chromosome's fitness values (valid
 fitness values are always positive).
<P>
<DL>
<DT><B>Since:</B></DT>
  <DD>2.0 (until 1.1: type int)</DD>
</DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="Chromosome()"><!-- --></A><H3>
Chromosome</H3>
<PRE>
public <B>Chromosome</B>()
           throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DD>Default constructor, only provided for dynamic instantiation.<p>
 Attention: The configuration used is the one set with the static method
 Genotype.setConfiguration.
<P>
<DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE><DT><B>Since:</B></DT>
  <DD>2.4</DD>
</DL>
</DL>
<HR>

<A NAME="Chromosome(org.jgap.Configuration)"><!-- --></A><H3>
Chromosome</H3>
<PRE>
public <B>Chromosome</B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_configuration)
           throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DD>Constructor, provided for dynamic or minimal instantiation.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>a_configuration</CODE> - the configuration to use
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE><DT><B>Since:</B></DT>
  <DD>3.0</DD>
</DL>
</DL>
<HR>

<A NAME="Chromosome(org.jgap.Configuration, java.lang.String)"><!-- --></A><H3>
Chromosome</H3>
<PRE>
public <B>Chromosome</B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_configuration,
                  java.lang.String&nbsp;a_persistentRepresentatuion)
           throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A>,
                  <A HREF="../../org/jgap/UnsupportedRepresentationException.html" title="class in org.jgap">UnsupportedRepresentationException</A></PRE>
<DL>
<DD>Constructor, provided for instantiation via persistent representation.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>a_configuration</CODE> - the configuration to use<DD><CODE>a_persistentRepresentatuion</CODE> - valid persistent representation that
 was most likely obtained via getPersistentRepresentation()
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE>
<DD><CODE><A HREF="../../org/jgap/UnsupportedRepresentationException.html" title="class in org.jgap">UnsupportedRepresentationException</A></CODE><DT><B>Since:</B></DT>
  <DD>3.2</DD>
</DL>
</DL>
<HR>

<A NAME="Chromosome(org.jgap.Configuration, int)"><!-- --></A><H3>
Chromosome</H3>
<PRE>
public <B>Chromosome</B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_configuration,
                  int&nbsp;a_desiredSize)
           throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DD>Constructor for specifying the number of genes.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>a_configuration</CODE> - the configuration to use<DD><CODE>a_desiredSize</CODE> - number of genes the chromosome contains of
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE><DT><B>Since:</B></DT>
  <DD>2.2</DD>
</DL>
</DL>
<HR>

<A NAME="Chromosome(org.jgap.Configuration, org.jgap.Gene, int)"><!-- --></A><H3>
Chromosome</H3>
<PRE>
public <B>Chromosome</B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_configuration,
                  <A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>&nbsp;a_sampleGene,
                  int&nbsp;a_desiredSize)
           throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DD>Constructs a Chromosome of the given size separate from any specific
 Configuration. This constructor will use the given sample Gene to
 construct a new Chromosome instance containing genes all of the same
 type as the sample Gene. This can be useful for constructing sample
 chromosomes that use the same Gene type for all of their genes and that
 are to be used to setup a Configuration object.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>a_configuration</CODE> - the configuration to use<DD><CODE>a_sampleGene</CODE> - a concrete sampleGene instance that will be used
 as a template for all of the genes in this Chromosome<DD><CODE>a_desiredSize</CODE> - the desired size (number of genes) of this Chromosome
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE><DT><B>Since:</B></DT>
  <DD>1.0</DD>
</DL>
</DL>
<HR>

<A NAME="Chromosome(org.jgap.Configuration, org.jgap.Gene, int, org.jgap.IGeneConstraintChecker)"><!-- --></A><H3>
Chromosome</H3>
<PRE>
public <B>Chromosome</B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_configuration,
                  <A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>&nbsp;a_sampleGene,
                  int&nbsp;a_desiredSize,
                  <A HREF="../../org/jgap/IGeneConstraintChecker.html" title="interface in org.jgap">IGeneConstraintChecker</A>&nbsp;a_constraintChecker)
           throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE></DL>
</DL>
<HR>

<A NAME="Chromosome(org.jgap.Configuration, org.jgap.Gene[])"><!-- --></A><H3>
Chromosome</H3>
<PRE>
public <B>Chromosome</B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_configuration,
                  <A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>[]&nbsp;a_initialGenes)
           throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DD>Constructs a Chromosome separate from any specific Configuration. This
 can be useful for constructing sample chromosomes that are to be used
 to setup a Configuration object.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>a_configuration</CODE> - the configuration to use<DD><CODE>a_initialGenes</CODE> - the initial genes of this Chromosome
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE><DT><B>Since:</B></DT>
  <DD>1.0</DD>
</DL>
</DL>
<HR>

<A NAME="Chromosome(org.jgap.Configuration, org.jgap.Gene[], org.jgap.IGeneConstraintChecker)"><!-- --></A><H3>
Chromosome</H3>
<PRE>
public <B>Chromosome</B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_configuration,
                  <A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>[]&nbsp;a_initialGenes,
                  <A HREF="../../org/jgap/IGeneConstraintChecker.html" title="interface in org.jgap">IGeneConstraintChecker</A>&nbsp;a_constraintChecker)
           throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DD>Constructs a Chromosome separate from any specific Configuration. This
 can be useful for constructing sample chromosomes that are to be used
 to setup a Configuration object. Additionally, a constraint checker can be
 specified. It is used right here to verify the validity of the gene types
 supplied.
<P>
<DL>
<DT><B>Parameters:</B><DD><CODE>a_configuration</CODE> - the configuration to use<DD><CODE>a_initialGenes</CODE> - the initial genes of this Chromosome<DD><CODE>a_constraintChecker</CODE> - constraint checker to use
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE> - in case the constraint checker
 reports a configuration error<DT><B>Since:</B></DT>
  <DD>2.5</DD>
</DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="initFromGene(org.jgap.Gene)"><!-- --></A><H3>
initFromGene</H3>
<PRE>
protected void <B>initFromGene</B>(<A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>&nbsp;a_sampleGene)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="checkGenes(org.jgap.Gene[])"><!-- --></A><H3>
checkGenes</H3>
<PRE>
protected void <B>checkGenes</B>(<A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>[]&nbsp;a_initialGenes)</PRE>
<DL>
<DD>Helper: called by constructors only to verify the initial genes.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_initialGenes</CODE> - the initial genes of this Chromosome to verify<DT><B>Since:</B></DT>
  <DD>2.5</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="clone()"><!-- --></A><H3>
clone</H3>
<PRE>
public java.lang.Object <B>clone</B>()</PRE>
<DL>
<DD>Returns a copy of this Chromosome. The returned instance can evolve
 independently of this instance. Note that, if possible, this method
 will first attempt to acquire a Chromosome instance from the active
 ChromosomePool (if any) and set its value appropriately before
 returning it. If that is not possible, then a new Chromosome instance
 will be constructed and its value set appropriately before returning.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../org/jgap/util/ICloneable.html#clone()">clone</A></CODE> in interface <CODE><A HREF="../../org/jgap/util/ICloneable.html" title="interface in org.jgap.util">ICloneable</A></CODE><DT><B>Specified by:</B><DD><CODE><A HREF="../../org/jgap/BaseChromosome.html#clone()">clone</A></CODE> in class <CODE><A HREF="../../org/jgap/BaseChromosome.html" title="class in org.jgap">BaseChromosome</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>copy of this Chromosome
<DT><B>Throws:</B>
<DD><CODE>java.lang.IllegalStateException</CODE> - instead of CloneNotSupportedException<DT><B>Since:</B></DT>
  <DD>1.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="cloneObject(java.lang.Object)"><!-- --></A><H3>
cloneObject</H3>
<PRE>
protected java.lang.Object <B>cloneObject</B>(java.lang.Object&nbsp;a_object)
                                throws java.lang.Exception</PRE>
<DL>
<DD>Clones an object by using clone handlers. If no deep cloning possible, then
 return the reference.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_object</CODE> - the object to clone
<DT><B>Returns:</B><DD>the cloned object, or the object itself if no coning supported
<DT><B>Throws:</B>
<DD><CODE>java.lang.Exception</CODE><DT><B>Since:</B></DT>
  <DD>2.6</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="getFitnessValue()"><!-- --></A><H3>
getFitnessValue</H3>
<PRE>
public double <B>getFitnessValue</B>()</PRE>
<DL>
<DD>Retrieves the fitness value of this Chromosome, as determined by the
 active fitness function. If a bulk fitness function is in use and
 has not yet assigned a fitness value to this Chromosome, then -1 is
 returned.<p>
 Attention: should not be called from toString() as the fitness value would
 be computed if it was initial!
<P>
<DD><DL>

<DT><B>Returns:</B><DD>a positive double value representing the fitness of this
 Chromosome, or -1 if a bulk fitness function is in use and has not yet
 assigned a fitness value to this Chromosome<DT><B>Since:</B></DT>
  <DD>2.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="getFitnessValueDirectly()"><!-- --></A><H3>
getFitnessValueDirectly</H3>
<PRE>
public double <B>getFitnessValueDirectly</B>()</PRE>
<DL>
<DD><DL>

<DT><B>Returns:</B><DD>the lastly computed fitness value, or FitnessFunction.NO_FITNESS_VALUE
 in case no value has been computed yet.</DL>
</DD>
</DL>
<HR>

<A NAME="calcFitnessValue()"><!-- --></A><H3>
calcFitnessValue</H3>
<PRE>
protected double <B>calcFitnessValue</B>()</PRE>
<DL>
<DD><DL>

<DT><B>Returns:</B><DD>fitness value of this chromosome determined via the registered
 fitness function<DT><B>Since:</B></DT>
  <DD>2.4</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="setFitnessValue(double)"><!-- --></A><H3>
setFitnessValue</H3>
<PRE>
public void <B>setFitnessValue</B>(double&nbsp;a_newFitnessValue)</PRE>
<DL>
<DD>Sets the fitness value of this Chromosome. This method is for use
 by bulk fitness functions and should not be invokved from anything
 else (except test cases).
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_newFitnessValue</CODE> - a positive integer representing the fitness
 of this Chromosome<DT><B>Since:</B></DT>
  <DD>1.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="setFitnessValueDirectly(double)"><!-- --></A><H3>
setFitnessValueDirectly</H3>
<PRE>
public void <B>setFitnessValueDirectly</B>(double&nbsp;a_newFitnessValue)</PRE>
<DL>
<DD>Sets the fitness value of this Chromosome directly without any
 constraint checks, conversions or checks. Only use if you know what
 you do.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_newFitnessValue</CODE> - a positive integer representing the fitness
 of this Chromosome</DL>
</DD>
</DL>
<HR>

<A NAME="toString()"><!-- --></A><H3>
toString</H3>
<PRE>
public java.lang.String <B>toString</B>()</PRE>
<DL>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE>toString</CODE> in class <CODE>java.lang.Object</CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>a string representation of this Chromosome, useful for display
 purposes<DT><B>Since:</B></DT>
  <DD>1.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="randomInitialChromosome(org.jgap.Configuration)"><!-- --></A><H3>
randomInitialChromosome</H3>
<PRE>
public static <A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A> <B>randomInitialChromosome</B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_configuration)
                                           throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DD>Convenience method that returns a new Chromosome instance with its
 genes values (alleles) randomized. Note that, if possible, this method
 will acquire a Chromosome instance from the active ChromosomePool
 (if any) and then randomize its gene values before returning it. If a
 Chromosome cannot be acquired from the pool, then a new instance will
 be constructed and its gene values randomized before returning it.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_configuration</CODE> - the configuration to use
<DT><B>Returns:</B><DD>randomly initialized Chromosome
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE> - if the given Configuration
 instance is invalid
<DD><CODE>java.lang.IllegalArgumentException</CODE> - if the given Configuration instance
 is null<DT><B>Since:</B></DT>
  <DD>1.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="equals(java.lang.Object)"><!-- --></A><H3>
equals</H3>
<PRE>
public boolean <B>equals</B>(java.lang.Object&nbsp;other)</PRE>
<DL>
<DD>Compares this Chromosome against the specified object. The result is
 true if and the argument is an instance of the Chromosome class
 and has a set of genes equal to this one.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE>equals</CODE> in class <CODE>java.lang.Object</CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>other</CODE> - the object to compare against
<DT><B>Returns:</B><DD>true: if the objects are the same, false otherwise<DT><B>Since:</B></DT>
  <DD>1.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="hashCode()"><!-- --></A><H3>
hashCode</H3>
<PRE>
public int <B>hashCode</B>()</PRE>
<DL>
<DD>Retrieve a hash code for this Chromosome. Does not considers the order
 of the Genes for all cases (especially when gene is empty).
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE>hashCode</CODE> in class <CODE>java.lang.Object</CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the hash code of this Chromosome<DT><B>Since:</B></DT>
  <DD>1.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="compareTo(java.lang.Object)"><!-- --></A><H3>
compareTo</H3>
<PRE>
public int <B>compareTo</B>(java.lang.Object&nbsp;other)</PRE>
<DL>
<DD>Compares the given Chromosome to this Chromosome. This chromosome is
 considered to be "less than" the given chromosome if it has a fewer
 number of genes or if any of its gene values (alleles) are less than
 their corresponding gene values in the other chromosome.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>other</CODE> - the Chromosome against which to compare this chromosome
<DT><B>Returns:</B><DD>a negative number if this chromosome is "less than" the given
 chromosome, zero if they are equal to each other, and a positive number if
 this chromosome is "greater than" the given chromosome<DT><B>Since:</B></DT>
  <DD>1.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="setIsSelectedForNextGeneration(boolean)"><!-- --></A><H3>
setIsSelectedForNextGeneration</H3>
<PRE>
public void <B>setIsSelectedForNextGeneration</B>(boolean&nbsp;a_isSelected)</PRE>
<DL>
<DD>Sets whether this Chromosome has been selected by the natural selector
 to continue to the next generation or manually (e.g. via an add-method).
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_isSelected</CODE> - true if this Chromosome has been selected, false
 otherwise<DT><B>Since:</B></DT>
  <DD>1.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="isSelectedForNextGeneration()"><!-- --></A><H3>
isSelectedForNextGeneration</H3>
<PRE>
public boolean <B>isSelectedForNextGeneration</B>()</PRE>
<DL>
<DD>Retrieves whether this Chromosome has been selected by the natural
 selector to continue to the next generation.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>true if this Chromosome has been selected, false otherwise<DT><B>Since:</B></DT>
  <DD>1.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="cleanup()"><!-- --></A><H3>
cleanup</H3>
<PRE>
public void <B>cleanup</B>()</PRE>
<DL>
<DD>Invoked when this Chromosome is no longer needed and should perform
 any necessary cleanup. Note that this method will attempt to release
 this Chromosome instance to the active ChromosomePool, if any.
<P>
<DD><DL>
<DT><B>Since:</B></DT>
  <DD>1.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="setApplicationData(java.lang.Object)"><!-- --></A><H3>
setApplicationData</H3>
<PRE>
public void <B>setApplicationData</B>(java.lang.Object&nbsp;a_newData)</PRE>
<DL>
<DD>This sets the application-specific data that is attached to this
 Chromosome. Attaching application-specific data may be useful for
 some applications when it comes time to evaluate this Chromosome
 in the fitness function. JGAP ignores this data, except for cloning and
 comparison (latter only if opted in via setCompareApplicationData(..))
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_newData</CODE> - the new application-specific data to attach to this
 Chromosome. Should be an instance of IApplicationData<DT><B>Since:</B></DT>
  <DD>1.1</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="getApplicationData()"><!-- --></A><H3>
getApplicationData</H3>
<PRE>
public java.lang.Object <B>getApplicationData</B>()</PRE>
<DL>
<DD>Retrieves the application-specific data that is attached to this
 Chromosome. Attaching application-specific data may be useful for
 some applications when it comes time to evaluate this Chromosome
 in the fitness function. JGAP ignores this data functionally.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>the application-specific data previously attached to this
 Chromosome, or null if there is no data attached<DT><B>Since:</B></DT>
  <DD>1.1</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="setGenes(org.jgap.Gene[])"><!-- --></A><H3>
setGenes</H3>
<PRE>
public void <B>setGenes</B>(<A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>[]&nbsp;a_genes)
              throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DD>Sets the genes for the chromosome.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../org/jgap/IChromosome.html#setGenes(org.jgap.Gene[])">setGenes</A></CODE> in interface <CODE><A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../org/jgap/BaseChromosome.html#setGenes(org.jgap.Gene[])">setGenes</A></CODE> in class <CODE><A HREF="../../org/jgap/BaseChromosome.html" title="class in org.jgap">BaseChromosome</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_genes</CODE> - the genes to set for the chromosome
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE> - in case constraint checker is
 provided</DL>
</DD>
</DL>
<HR>

<A NAME="setCompareApplicationData(boolean)"><!-- --></A><H3>
setCompareApplicationData</H3>
<PRE>
public void <B>setCompareApplicationData</B>(boolean&nbsp;a_doCompare)</PRE>
<DL>
<DD>Should we also consider the application data when comparing? Default is
 "false" as "true" means a Chromosome is losing its identity when
 application data is set differently!
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_doCompare</CODE> - true: consider application data in method compareTo<DT><B>Since:</B></DT>
  <DD>2.2</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="isCompareApplicationData()"><!-- --></A><H3>
isCompareApplicationData</H3>
<PRE>
public boolean <B>isCompareApplicationData</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="setConstraintChecker(org.jgap.IGeneConstraintChecker)"><!-- --></A><H3>
setConstraintChecker</H3>
<PRE>
public void <B>setConstraintChecker</B>(<A HREF="../../org/jgap/IGeneConstraintChecker.html" title="interface in org.jgap">IGeneConstraintChecker</A>&nbsp;a_constraintChecker)
                          throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DD>Sets the constraint checker to be used for this gene whenever method
 setAllele(Object) is called.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_constraintChecker</CODE> - the constraint checker to be set
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE><DT><B>Since:</B></DT>
  <DD>2.5</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="getConstraintChecker()"><!-- --></A><H3>
getConstraintChecker</H3>
<PRE>
public <A HREF="../../org/jgap/IGeneConstraintChecker.html" title="interface in org.jgap">IGeneConstraintChecker</A> <B>getConstraintChecker</B>()</PRE>
<DL>
<DD><DL>

<DT><B>Returns:</B><DD>IGeneConstraintChecker the constraint checker to be used whenever
 method setGenes(Gene[]) is called.<DT><B>Since:</B></DT>
  <DD>2.5</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="verify(org.jgap.IGeneConstraintChecker)"><!-- --></A><H3>
verify</H3>
<PRE>
protected void <B>verify</B>(<A HREF="../../org/jgap/IGeneConstraintChecker.html" title="interface in org.jgap">IGeneConstraintChecker</A>&nbsp;a_constraintChecker)
               throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DD>Verifies the state of the chromosome. Especially takes care of the
 given constraint checker.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_constraintChecker</CODE> - the constraint checker to verify
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE><DT><B>Since:</B></DT>
  <DD>2.5</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="isHandlerFor(java.lang.Object, java.lang.Class)"><!-- --></A><H3>
isHandlerFor</H3>
<PRE>
public boolean <B>isHandlerFor</B>(java.lang.Object&nbsp;a_obj,
                            java.lang.Class&nbsp;a_class)</PRE>
<DL>
<DD>Determines whether the handler is suitable for the given object instance
 or class.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_obj</CODE> - the object instance to check<DD><CODE>a_class</CODE> - the class to check, alternatively if no object instance is
 available (some handlers could need to get an object and would always
 return false if only a class is provided)
<DT><B>Returns:</B><DD>true: handler suitable for given object or class</DL>
</DD>
</DL>
<HR>

<A NAME="perform(java.lang.Object, java.lang.Class, java.lang.Object)"><!-- --></A><H3>
perform</H3>
<PRE>
public java.lang.Object <B>perform</B>(java.lang.Object&nbsp;a_obj,
                                java.lang.Class&nbsp;a_class,
                                java.lang.Object&nbsp;a_params)
                         throws java.lang.Exception</PRE>
<DL>
<DD>Performs a task for the given object or class. For some handlers,
 additional parameters are necessary, which could be provided.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_obj</CODE> - the object instance to perform the handler task for<DD><CODE>a_class</CODE> - the class to perform the handler task for, in case no object
 instance is available (some handlers could need to get an object and would
 always throw an exception if only a class is provided)<DD><CODE>a_params</CODE> - optional parameters needed for the handler to perform its
 task
<DT><B>Returns:</B><DD>Object
<DT><B>Throws:</B>
<DD><CODE>java.lang.Exception</CODE> - in case of any error</DL>
</DD>
</DL>
<HR>

<A NAME="setMultiObjectives(java.util.List)"><!-- --></A><H3>
setMultiObjectives</H3>
<PRE>
public void <B>setMultiObjectives</B>(java.util.List&nbsp;a_values)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="getMultiObjectives()"><!-- --></A><H3>
getMultiObjectives</H3>
<PRE>
public java.util.List <B>getMultiObjectives</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
JGAP</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../org/jgap/CachedFitnessFunction.html" title="class in org.jgap"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?org/jgap/Chromosome.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Chromosome.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
