<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_02) on Mon Nov 05 17:25:02 CET 2007 -->
<TITLE>
Population
</TITLE>

<META NAME="date" CONTENT="2007-11-05">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="Population";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
JGAP</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../org/jgap/NaturalSelector.FitnessValueComparator.html" title="class in org.jgap"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../org/jgap/RandomGenerator.html" title="interface in org.jgap"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?org/jgap/Population.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Population.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.jgap</FONT>
<BR>
Class Population</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../resources/inherit.gif" ALT="extended by "><B>org.jgap.Population</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD>java.io.Serializable, java.lang.Cloneable, <A HREF="../../org/jgap/IPersistentRepresentation.html" title="interface in org.jgap">IPersistentRepresentation</A>, <A HREF="../../org/jgap/util/ICloneable.html" title="interface in org.jgap.util">ICloneable</A></DD>
</DL>
<HR>
<DL>
<DT><PRE>public class <B>Population</B><DT>extends java.lang.Object<DT>implements java.io.Serializable, <A HREF="../../org/jgap/util/ICloneable.html" title="interface in org.jgap.util">ICloneable</A>, <A HREF="../../org/jgap/IPersistentRepresentation.html" title="interface in org.jgap">IPersistentRepresentation</A></DL>
</PRE>

<P>
List of chromosomes held in the Genotype (or possibly later in the
 Configuration object).
<P>

<P>
<DL>
<DT><B>Since:</B></DT>
  <DD>2.0</DD>
<DT><B>See Also:</B><DD><A HREF="../../serialized-form.html#org.jgap.Population">Serialized Form</A></DL>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#CHROM_DELIMITER">CHROM_DELIMITER</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#CHROM_DELIMITER_CLOSING">CHROM_DELIMITER_CLOSING</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Represents the closing delimiter that is used to separate chromosomes in
 the persistent representation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#CHROM_DELIMITER_HEADING">CHROM_DELIMITER_HEADING</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Represents the heading delimiter that is used to separate chromosomes in
 the persistent representation.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../org/jgap/Population.html#Population()">Population</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../org/jgap/Population.html#Population(org.jgap.Configuration)">Population</A></B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_config)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../org/jgap/Population.html#Population(org.jgap.Configuration, org.jgap.IChromosome)">Population</A></B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_config,
           <A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A>&nbsp;a_chromosome)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../org/jgap/Population.html#Population(org.jgap.Configuration, org.jgap.IChromosome[])">Population</A></B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_config,
           <A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A>[]&nbsp;a_chromosomes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../org/jgap/Population.html#Population(org.jgap.Configuration, int)">Population</A></B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_config,
           int&nbsp;a_size)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#addChromosome(org.jgap.IChromosome)">addChromosome</A></B>(<A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A>&nbsp;a_toAdd)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Adds a Chromosome to this Population.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#addChromosomes(org.jgap.Population)">addChromosomes</A></B>(<A HREF="../../org/jgap/Population.html" title="class in org.jgap">Population</A>&nbsp;a_population)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Adds all the Chromosomes in the given Population.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#clear()">clear</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Clears the list of chromosomes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#clone()">clone</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#compareTo(java.lang.Object)">compareTo</A></B>(java.lang.Object&nbsp;a_pop)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;This method is not producing symmetric results as -1 is more often returned
 than 1 (see description of return value).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#contains(org.jgap.IChromosome)">contains</A></B>(<A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A>&nbsp;a_chromosome)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Determines whether the given chromosome is contained within the population.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#createChromosome(java.lang.String, java.lang.String)">createChromosome</A></B>(java.lang.String&nbsp;a_chromClassName,
                 java.lang.String&nbsp;a_persistentRepresentation)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Creates a new Chromosome instance.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#determineFittestChromosome()">determineFittestChromosome</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Determines the fittest Chromosome in the Population (the one with the
 highest fitness value) and memorizes it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#determineFittestChromosome(int, int)">determineFittestChromosome</A></B>(int&nbsp;a_startIndex,
                           int&nbsp;a_endIndex)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Determines the fittest Chromosome in the population (the one with the
 highest fitness value) within the given indices and memorizes it.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#determineFittestChromosomes(int)">determineFittestChromosomes</A></B>(int&nbsp;a_numberOfChromosomes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sorts the Chromosome list and returns the fittest n Chromosomes in
 the population.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#equals(java.lang.Object)">equals</A></B>(java.lang.Object&nbsp;a_pop)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The equals-method.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#getChromosome(int)">getChromosome</A></B>(int&nbsp;a_index)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#getChromosomes()">getChromosomes</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#getConfiguration()">getConfiguration</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#getGenome(boolean)">getGenome</A></B>(boolean&nbsp;a_resolveCompositeGenes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns the genotype of the population, i.e. the list of genes in the
 Population.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#getPersistentRepresentation()">getPersistentRepresentation</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Returns a persistent representation of this chromosome, see interface Gene
 for description.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#isChanged()">isChanged</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#isSorted()">isSorted</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.util.Iterator</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#iterator()">iterator</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#keepPopSizeConstant()">keepPopSizeConstant</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Cares that the population size does not exceed the maximum size given in
 the configuration.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#setChanged(boolean)">setChanged</A></B>(boolean&nbsp;a_changed)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Mark that for the population the fittest chromosome may have changed.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#setChromosome(int, org.jgap.IChromosome)">setChromosome</A></B>(int&nbsp;a_index,
              <A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A>&nbsp;a_chromosome)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets in the given Chromosome on the given index in the list of chromosomes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#setChromosomes(java.util.List)">setChromosomes</A></B>(java.util.List&nbsp;a_chromosomes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Replaces all chromosomes in the population with the give list of
 chromosomes.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#setSorted(boolean)">setSorted</A></B>(boolean&nbsp;a_sorted)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Mark the population as sorted.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#setValueFromPersistentRepresentation(java.lang.String)">setValueFromPersistentRepresentation</A></B>(java.lang.String&nbsp;a_representation)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Counterpart of getPersistentRepresentation.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#size()">size</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#sort(java.util.Comparator)">sort</A></B>(java.util.Comparator&nbsp;a_comparator)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sorts the chromosomes within the population utilzing the given comparator.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#sortByFitness()">sortByFitness</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sorts the chromosomes within the population according to their fitness
 value using ChromosomFitnessComparator.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected static&nbsp;java.util.List</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#split(java.lang.String)">split</A></B>(java.lang.String&nbsp;a_string)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Splits the input a_string into individual chromosome representations.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A>[]</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../org/jgap/Population.html#toChromosomes()">toChromosomes</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="CHROM_DELIMITER"><!-- --></A><H3>
CHROM_DELIMITER</H3>
<PRE>
public static final java.lang.String <B>CHROM_DELIMITER</B></PRE>
<DL>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#org.jgap.Population.CHROM_DELIMITER">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="CHROM_DELIMITER_HEADING"><!-- --></A><H3>
CHROM_DELIMITER_HEADING</H3>
<PRE>
public static final java.lang.String <B>CHROM_DELIMITER_HEADING</B></PRE>
<DL>
<DD>Represents the heading delimiter that is used to separate chromosomes in
 the persistent representation.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#org.jgap.Population.CHROM_DELIMITER_HEADING">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="CHROM_DELIMITER_CLOSING"><!-- --></A><H3>
CHROM_DELIMITER_CLOSING</H3>
<PRE>
public static final java.lang.String <B>CHROM_DELIMITER_CLOSING</B></PRE>
<DL>
<DD>Represents the closing delimiter that is used to separate chromosomes in
 the persistent representation.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#org.jgap.Population.CHROM_DELIMITER_CLOSING">Constant Field Values</A></DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="Population(org.jgap.Configuration)"><!-- --></A><H3>
Population</H3>
<PRE>
public <B>Population</B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_config)
           throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE></DL>
</DL>
<HR>

<A NAME="Population(org.jgap.Configuration, org.jgap.IChromosome[])"><!-- --></A><H3>
Population</H3>
<PRE>
public <B>Population</B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_config,
                  <A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A>[]&nbsp;a_chromosomes)
           throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE></DL>
</DL>
<HR>

<A NAME="Population(org.jgap.Configuration, org.jgap.IChromosome)"><!-- --></A><H3>
Population</H3>
<PRE>
public <B>Population</B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_config,
                  <A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A>&nbsp;a_chromosome)
           throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE></DL>
</DL>
<HR>

<A NAME="Population(org.jgap.Configuration, int)"><!-- --></A><H3>
Population</H3>
<PRE>
public <B>Population</B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_config,
                  int&nbsp;a_size)
           throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE></DL>
</DL>
<HR>

<A NAME="Population()"><!-- --></A><H3>
Population</H3>
<PRE>
public <B>Population</B>()
           throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE></DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="getConfiguration()"><!-- --></A><H3>
getConfiguration</H3>
<PRE>
public <A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A> <B>getConfiguration</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="addChromosome(org.jgap.IChromosome)"><!-- --></A><H3>
addChromosome</H3>
<PRE>
public void <B>addChromosome</B>(<A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A>&nbsp;a_toAdd)</PRE>
<DL>
<DD>Adds a Chromosome to this Population. Does nothing when given null.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_toAdd</CODE> - the Chromosome to add<DT><B>Since:</B></DT>
  <DD>2.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="addChromosomes(org.jgap.Population)"><!-- --></A><H3>
addChromosomes</H3>
<PRE>
public void <B>addChromosomes</B>(<A HREF="../../org/jgap/Population.html" title="class in org.jgap">Population</A>&nbsp;a_population)</PRE>
<DL>
<DD>Adds all the Chromosomes in the given Population. Does nothing on null or
 an empty Population.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_population</CODE> - the Population to add<DT><B>Since:</B></DT>
  <DD>2.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="setChromosomes(java.util.List)"><!-- --></A><H3>
setChromosomes</H3>
<PRE>
public void <B>setChromosomes</B>(java.util.List&nbsp;a_chromosomes)</PRE>
<DL>
<DD>Replaces all chromosomes in the population with the give list of
 chromosomes.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_chromosomes</CODE> - the chromosomes to make the population up from</DL>
</DD>
</DL>
<HR>

<A NAME="setChromosome(int, org.jgap.IChromosome)"><!-- --></A><H3>
setChromosome</H3>
<PRE>
public void <B>setChromosome</B>(int&nbsp;a_index,
                          <A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A>&nbsp;a_chromosome)</PRE>
<DL>
<DD>Sets in the given Chromosome on the given index in the list of chromosomes.
 If the given index is exceeding the list by one, the chromosome is
 appended.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_index</CODE> - the index to set the Chromosome in<DD><CODE>a_chromosome</CODE> - the Chromosome to be set<DT><B>Since:</B></DT>
  <DD>2.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="getChromosomes()"><!-- --></A><H3>
getChromosomes</H3>
<PRE>
public java.util.List <B>getChromosomes</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the list of Chromosome's in the Population. Don't modify the
 retrieved list by using clear(), remove(int) etc. If you do so, you need to
 call setChanged(true)<DT><B>Since:</B></DT>
  <DD>2.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="getChromosome(int)"><!-- --></A><H3>
getChromosome</H3>
<PRE>
public <A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A> <B>getChromosome</B>(int&nbsp;a_index)</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_index</CODE> - the index of the Chromosome to be returned
<DT><B>Returns:</B><DD>Chromosome at given index in the Population<DT><B>Since:</B></DT>
  <DD>2.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="size()"><!-- --></A><H3>
size</H3>
<PRE>
public int <B>size</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>number of Chromosome's in the Population<DT><B>Since:</B></DT>
  <DD>2.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="iterator()"><!-- --></A><H3>
iterator</H3>
<PRE>
public java.util.Iterator <B>iterator</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>Iterator for the Chromosome list in the Population. Please be aware
 that using remove() forces you to call setChanged(true)<DT><B>Since:</B></DT>
  <DD>2.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="toChromosomes()"><!-- --></A><H3>
toChromosomes</H3>
<PRE>
public <A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A>[] <B>toChromosomes</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the Population converted into a list of Chromosome's<DT><B>Since:</B></DT>
  <DD>2.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="determineFittestChromosome()"><!-- --></A><H3>
determineFittestChromosome</H3>
<PRE>
public <A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A> <B>determineFittestChromosome</B>()</PRE>
<DL>
<DD>Determines the fittest Chromosome in the Population (the one with the
 highest fitness value) and memorizes it. This is an optimized version
 compared to calling determineFittesChromosomes(1).
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>the fittest Chromosome of the Population<DT><B>Since:</B></DT>
  <DD>2.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="determineFittestChromosome(int, int)"><!-- --></A><H3>
determineFittestChromosome</H3>
<PRE>
public <A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A> <B>determineFittestChromosome</B>(int&nbsp;a_startIndex,
                                              int&nbsp;a_endIndex)</PRE>
<DL>
<DD>Determines the fittest Chromosome in the population (the one with the
 highest fitness value) within the given indices and memorizes it. This is
 an optimized version compared to calling determineFittesChromosomes(1).
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_startIndex</CODE> - index to begin the evaluation with<DD><CODE>a_endIndex</CODE> - index to end the evaluation with
<DT><B>Returns:</B><DD>the fittest Chromosome of the population within the given indices<DT><B>Since:</B></DT>
  <DD>3.0</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="setChanged(boolean)"><!-- --></A><H3>
setChanged</H3>
<PRE>
protected void <B>setChanged</B>(boolean&nbsp;a_changed)</PRE>
<DL>
<DD>Mark that for the population the fittest chromosome may have changed.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_changed</CODE> - true: population's fittest chromosome may have changed,
 false: fittest chromosome evaluated earlier is still valid<DT><B>Since:</B></DT>
  <DD>2.2</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="isChanged()"><!-- --></A><H3>
isChanged</H3>
<PRE>
public boolean <B>isChanged</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>true: population's chromosomes (maybe) were changed,
 false: not changed for sure<DT><B>Since:</B></DT>
  <DD>2.6</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="setSorted(boolean)"><!-- --></A><H3>
setSorted</H3>
<PRE>
protected void <B>setSorted</B>(boolean&nbsp;a_sorted)</PRE>
<DL>
<DD>Mark the population as sorted.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_sorted</CODE> - true: mark population as sorted<DT><B>Since:</B></DT>
  <DD>2.6</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="contains(org.jgap.IChromosome)"><!-- --></A><H3>
contains</H3>
<PRE>
public boolean <B>contains</B>(<A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A>&nbsp;a_chromosome)</PRE>
<DL>
<DD>Determines whether the given chromosome is contained within the population.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_chromosome</CODE> - the chromosome to check
<DT><B>Returns:</B><DD>true: chromosome contained within population<DT><B>Since:</B></DT>
  <DD>2.1</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="keepPopSizeConstant()"><!-- --></A><H3>
keepPopSizeConstant</H3>
<PRE>
public void <B>keepPopSizeConstant</B>()
                         throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DD>Cares that the population size does not exceed the maximum size given in
 the configuration.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>

<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE><DT><B>Since:</B></DT>
  <DD>3.2</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="determineFittestChromosomes(int)"><!-- --></A><H3>
determineFittestChromosomes</H3>
<PRE>
public java.util.List <B>determineFittestChromosomes</B>(int&nbsp;a_numberOfChromosomes)</PRE>
<DL>
<DD>Sorts the Chromosome list and returns the fittest n Chromosomes in
 the population.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_numberOfChromosomes</CODE> - number of top performer chromosomes to be
 returned
<DT><B>Returns:</B><DD>list of the fittest n Chromosomes of the population, or the fittest
 x Chromosomes with x = number of chromosomes in case n > x.<DT><B>Since:</B></DT>
  <DD>2.4</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="sortByFitness()"><!-- --></A><H3>
sortByFitness</H3>
<PRE>
public void <B>sortByFitness</B>()</PRE>
<DL>
<DD>Sorts the chromosomes within the population according to their fitness
 value using ChromosomFitnessComparator. The fittest chromosome is then
 at index 0.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Since:</B></DT>
  <DD>2.6</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="sort(java.util.Comparator)"><!-- --></A><H3>
sort</H3>
<PRE>
protected void <B>sort</B>(java.util.Comparator&nbsp;a_comparator)</PRE>
<DL>
<DD>Sorts the chromosomes within the population utilzing the given comparator.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_comparator</CODE> - the comparator to utilize for sorting<DT><B>Since:</B></DT>
  <DD>2.6</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="getGenome(boolean)"><!-- --></A><H3>
getGenome</H3>
<PRE>
public java.util.List <B>getGenome</B>(boolean&nbsp;a_resolveCompositeGenes)</PRE>
<DL>
<DD>Returns the genotype of the population, i.e. the list of genes in the
 Population.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_resolveCompositeGenes</CODE> - true: split encountered CompositeGenes
 into their single (atomic) genes
<DT><B>Returns:</B><DD>genotype of the population<DT><B>Since:</B></DT>
  <DD>2.3</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="isSorted()"><!-- --></A><H3>
isSorted</H3>
<PRE>
public boolean <B>isSorted</B>()</PRE>
<DL>
<DD><DL>
</DL>
</DD>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="equals(java.lang.Object)"><!-- --></A><H3>
equals</H3>
<PRE>
public boolean <B>equals</B>(java.lang.Object&nbsp;a_pop)</PRE>
<DL>
<DD>The equals-method.
<P>
<DD><DL>
<DT><B>Overrides:</B><DD><CODE>equals</CODE> in class <CODE>java.lang.Object</CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_pop</CODE> - the population instance to compare with
<DT><B>Returns:</B><DD>true: given object equal to comparing one<DT><B>Since:</B></DT>
  <DD>2.6</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="compareTo(java.lang.Object)"><!-- --></A><H3>
compareTo</H3>
<PRE>
public int <B>compareTo</B>(java.lang.Object&nbsp;a_pop)</PRE>
<DL>
<DD>This method is not producing symmetric results as -1 is more often returned
 than 1 (see description of return value).
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_pop</CODE> - the other population to compare
<DT><B>Returns:</B><DD>1: a_pop is null or having fewer chromosomes or equal number
 of chromosomes but at least one not contained. 0: both populations
 containing exactly the same chromosomes. -1: this population contains fewer
 chromosomes than a_pop<DT><B>Since:</B></DT>
  <DD>2.6</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="clone()"><!-- --></A><H3>
clone</H3>
<PRE>
public java.lang.Object <B>clone</B>()</PRE>
<DL>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../org/jgap/util/ICloneable.html#clone()">clone</A></CODE> in interface <CODE><A HREF="../../org/jgap/util/ICloneable.html" title="interface in org.jgap.util">ICloneable</A></CODE><DT><B>Overrides:</B><DD><CODE>clone</CODE> in class <CODE>java.lang.Object</CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>deeply cloned population instance<DT><B>Since:</B></DT>
  <DD>3.2</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="clear()"><!-- --></A><H3>
clear</H3>
<PRE>
public void <B>clear</B>()</PRE>
<DL>
<DD>Clears the list of chromosomes. Normally, this should not be necessary.
 But especially in distributed computing, a fresh population has to be
 provided sometimes.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Since:</B></DT>
  <DD>3.2</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="getPersistentRepresentation()"><!-- --></A><H3>
getPersistentRepresentation</H3>
<PRE>
public java.lang.String <B>getPersistentRepresentation</B>()</PRE>
<DL>
<DD>Returns a persistent representation of this chromosome, see interface Gene
 for description. Similar to CompositeGene's routine. But does not include
 all information of the chromosome (yet).
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../org/jgap/IPersistentRepresentation.html#getPersistentRepresentation()">getPersistentRepresentation</A></CODE> in interface <CODE><A HREF="../../org/jgap/IPersistentRepresentation.html" title="interface in org.jgap">IPersistentRepresentation</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>string representation of this Chromosome's relevant parts of its
 current state
<DT><B>Throws:</B>
<DD><CODE>java.lang.UnsupportedOperationException</CODE><DT><B>Since:</B></DT>
  <DD>3.2</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="setValueFromPersistentRepresentation(java.lang.String)"><!-- --></A><H3>
setValueFromPersistentRepresentation</H3>
<PRE>
public void <B>setValueFromPersistentRepresentation</B>(java.lang.String&nbsp;a_representation)
                                          throws <A HREF="../../org/jgap/UnsupportedRepresentationException.html" title="class in org.jgap">UnsupportedRepresentationException</A></PRE>
<DL>
<DD>Counterpart of getPersistentRepresentation.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../org/jgap/IPersistentRepresentation.html#setValueFromPersistentRepresentation(java.lang.String)">setValueFromPersistentRepresentation</A></CODE> in interface <CODE><A HREF="../../org/jgap/IPersistentRepresentation.html" title="interface in org.jgap">IPersistentRepresentation</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_representation</CODE> - the string representation retrieved from a prior
 call to the getPersistentRepresentation() method
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/UnsupportedRepresentationException.html" title="class in org.jgap">UnsupportedRepresentationException</A></CODE><DT><B>Since:</B></DT>
  <DD>3.2</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="createChromosome(java.lang.String, java.lang.String)"><!-- --></A><H3>
createChromosome</H3>
<PRE>
protected <A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A> <B>createChromosome</B>(java.lang.String&nbsp;a_chromClassName,
                                       java.lang.String&nbsp;a_persistentRepresentation)
                                throws java.lang.Exception</PRE>
<DL>
<DD>Creates a new Chromosome instance.<p>
 Taken and adapted from CompositeGene.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_chromClassName</CODE> - name of the Chromosome class<DD><CODE>a_persistentRepresentation</CODE> - persistent representation of the
 Chromosome to create (could be obtained via getPersistentRepresentation)
<DT><B>Returns:</B><DD>newly created Chromosome
<DT><B>Throws:</B>
<DD><CODE>java.lang.Exception</CODE><DT><B>Since:</B></DT>
  <DD>3.2</DD>
</DL>
</DD>
</DL>
<HR>

<A NAME="split(java.lang.String)"><!-- --></A><H3>
split</H3>
<PRE>
protected static final java.util.List <B>split</B>(java.lang.String&nbsp;a_string)
                                     throws <A HREF="../../org/jgap/UnsupportedRepresentationException.html" title="class in org.jgap">UnsupportedRepresentationException</A></PRE>
<DL>
<DD>Splits the input a_string into individual chromosome representations.<p>
 Taken and adapted from CompositeGene.
<P>
<DD><DL>
</DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_string</CODE> - the string to split
<DT><B>Returns:</B><DD>the elements of the returned array are the persistent
 representation strings of the population's components
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/UnsupportedRepresentationException.html" title="class in org.jgap">UnsupportedRepresentationException</A></CODE><DT><B>Since:</B></DT>
  <DD>3.2</DD>
</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
JGAP</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../org/jgap/NaturalSelector.FitnessValueComparator.html" title="class in org.jgap"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../org/jgap/RandomGenerator.html" title="interface in org.jgap"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?org/jgap/Population.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="Population.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
