<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_02) on Mon Nov 05 17:25:04 CET 2007 -->
<TITLE>
BulkFitnessOffsetRemover
</TITLE>

<META NAME="date" CONTENT="2007-11-05">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="BulkFitnessOffsetRemover";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
JGAP</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../org/jgap/impl/BooleanGene.html" title="class in org.jgap.impl"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../org/jgap/impl/CauchyRandomGenerator.html" title="class in org.jgap.impl"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html?org/jgap/impl/BulkFitnessOffsetRemover.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="BulkFitnessOffsetRemover.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
org.jgap.impl</FONT>
<BR>
Class BulkFitnessOffsetRemover</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../../resources/inherit.gif" ALT="extended by "><A HREF="../../../org/jgap/BulkFitnessFunction.html" title="class in org.jgap">org.jgap.BulkFitnessFunction</A>
      <IMG SRC="../../../resources/inherit.gif" ALT="extended by "><B>org.jgap.impl.BulkFitnessOffsetRemover</B>
</PRE>
<DL>
<DT><B>All Implemented Interfaces:</B> <DD>java.io.Serializable, java.lang.Cloneable, <A HREF="../../../org/jgap/util/ICloneable.html" title="interface in org.jgap.util">ICloneable</A></DD>
</DL>
<HR>
<DL>
<DT><PRE>public class <B>BulkFitnessOffsetRemover</B><DT>extends <A HREF="../../../org/jgap/BulkFitnessFunction.html" title="class in org.jgap">BulkFitnessFunction</A></DL>
</PRE>

<P>
Takes away the fitness offset of the population to evolve.
 The fitness function values of the population of <A HREF="../../../org/jgap/IChromosome.html" title="interface in org.jgap"><CODE>IChromosome</CODE></A>
 instances will start from a minimum of 1 afterwards.
 </p>
 <p>
 The removal of an offset in the fitness values of a population strengthens
 the "survival of the fittest" effect of a selector that performs selection
 upon fitness values. A high offset in the fitness values of a population
 lowers the relative difference between the fitness values of the Chromosomes
 in a population.
 </p>
 <h3>Example of applicability</h3>
 <p>
 You are optimizing a black box with <i>n</i> parameters that are mapped
 to <A HREF="../../../org/jgap/IChromosome.html" title="interface in org.jgap"><CODE>IChromosome</CODE></A> instances each having <i>n</i>
 <A HREF="../../../org/jgap/Gene.html" title="interface in org.jgap"><CODE>Gene</CODE></A> instances.<br>
 You want to minimize the answer time of the black box and provide
 a <A HREF="../../../org/jgap/FitnessFunction.html#evaluate(org.jgap.IChromosome)"><CODE>FitnessFunction.evaluate(org.jgap.IChromosome)</CODE></A>
 that takes the genes out of the chromosome, put's it's
 <A HREF="../../../org/jgap/Gene.html#getAllele()"><CODE>Gene.getAllele()</CODE></A>
 values to the parameters and measures the answer time of the black box
 (by invoking it's service to optimize). <br>
 The longer the time takes, the worse it's fitness is, so you have to
 invert the measured times to fitness values:
 <a name="bboptimizer"/>
 <pre>
 <font color="#0011EE">
 class BlackBoxOptimizer extends org.jgap.FitnessFunction{
   private BlackBox bbox;
   <font color="#999999">//Additional code: constructors</font>
   <font color="#999999">...</font>
   public double evaluate(org.jgap.IChromosome chromosome){
     double fitness = 0;
     <font color="#999999">// get the Gene[] & put the parameters into the box.
     ...
     </font>
     long duration = System.currentTimeMillis();
 <font color="#999999">
 // You certainly will use an advanced StopWatch...</font>
     this.bbox.service();
 <font color="#999999">// The black boxes service to optimize.</font>
     duration = System.currentTimeMillis()-duration;
     <font color="#999999">// transform the time into fitness value:</font>
     fitness = double.MAX_VALUE - (double)duration;
     return fitness;
   }
 }
 </font>
 </pre>
 </p>
 <p>
 <h4>We might get the following results (each row stands for a Chromosome,
 the table is a population):</h4>
 <table border="1">
 <tr align="left" valign="top">
 <th>
 duration
 </th>
 <th>
 fitness
 </th>
 <th>
 piece of fitness cake
 </th>
 </tr>
 <tr align="left" valign="top">
 <td>
 2000
 </td>
 <td>
 9218868437227403311
 </td>
 <td>
 33.333333333333336949106088992532 %
 </td>
 </tr>
 <tr align="left" valign="top">
 <td>
 3000
 </td>
 <td>
 9218868437227402311
 </td>
 <td>
 33.333333333333333333333333333333 %
 </td>
 </tr>
 <tr align="left" valign="top">
 <td>
 4000
 </td>
 <td>
 9218868437227401311
 </td>
 <td>
 33.333333333333329717560577674135 %
 </td>
 </tr>
 </table>
 </p>
 <p>
 If any <A HREF="../../../org/jgap/NaturalSelector.html" title="class in org.jgap"><CODE>NaturalSelector</CODE></A> performs selection based upon the
 fitness values, it will have to put those values in relation to each other.
 As a fact, the probability to select the Chromosome that contained the black
 box parameters that caused an answer time of 4000 ms is "equal" to the
 probability to select the Chromosome that caused a black box answer time to
 be 2000 ms!
 </p>
 <p>
 Of course one could work around that problem by replacing the
 <tt>Integer.MAX_VALUE</tt> transformation by a fixed maximum value the black
 box would need for the service.
 But what, if you have no guaranteed maximum answer time for the service of
 the black box ?
 Even if you have got one, it will be chosen sufficently high above the
 average answer time thus letting your fitness function return values with a
 high offset in the fitness.
 </p>
 <p>
 <h4>This is, what happens, if you use this instance for fitness
 evaluation:</h4>
 <table border="1">
 <tr align="left" valign="top">
 <th>
 duration
 </th>
 <th>
 fitness
 </th>
 <th>
 piece of fitness cake
 </th>
 </tr>
 <tr align="left" valign="top">
 <td>
 2000
 </td>
 <td>
 2001
 </td>
 <td>
 66.63 %
 </td>
 </tr>
 <tr align="left" valign="top">
 <td>
 3000
 </td>
 <td>
 1001
 </td>
 <td>
 33.33 %
 </td>
 </tr>
 <tr align="left" valign="top">
 <td>
 4000
 </td>
 <td>
 1
 </td>
 <td>
 0.03 %
 </td>
 </tr>
 </table>
 </p>
 <h3>Example of usage</h3>

 <p>
 This example shows how to use this instance for cutting fitness offsets.
 It is the same example as used <a href="#bboptimizer">above</a>.
 <pre>
 <font color="#0011EE">
 class BlackBoxOptimizer extends org.jgap.FitnessFunction{
   <font color="#999999">// Additional code: constructors
   ...</font>
   public double evaluate(org.jgap.IChromosome chromosome){
     <font color="#999999">.... // As shown above.</font>
   }

   public void startOptimization(org.jgap.Configuration gaConf)
       throws InvalidConfigurationException{
     <font color="#999999">
     // The given Configuration may be preconfigured with
     // NaturalSelector & GeneticOperator instances,.
     // But should not contain a FitnessFunction or BulkFitnessFunction!
 </font>
     <b>gaConf.setBulkFitnessFunction(new BulkFitnessOffsetRemover(this));</b>
     <font color="#999999">// Why does it work? We implement FitnessFunction!
     // Still to do here:
     // - Create a sample chromosome according to your blackbox & set it to
      //   the configuration.
     // - Create a random inital Genotype.
     // - loop over a desired amount of generations invoking
      //   aGenotype.evolve()..</font>
   }
 }
 </font>
 </pre>
 </p>
<P>

<P>
<DL>
<DT><B>Since:</B></DT>
  <DD>2.2</DD>
<DT><B>See Also:</B><DD><A HREF="../../../serialized-form.html#org.jgap.impl.BulkFitnessOffsetRemover">Serialized Form</A></DL>
<HR>

<P>

<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../../org/jgap/impl/BulkFitnessOffsetRemover.html#BulkFitnessOffsetRemover(org.jgap.FitnessFunction)">BulkFitnessOffsetRemover</A></B>(<A HREF="../../../org/jgap/FitnessFunction.html" title="class in org.jgap">FitnessFunction</A>&nbsp;a_ff)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.Object</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/impl/BulkFitnessOffsetRemover.html#clone()">clone</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Override in your implementation if necessary.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/impl/BulkFitnessOffsetRemover.html#evaluate(org.jgap.Population)">evaluate</A></B>(<A HREF="../../../org/jgap/Population.html" title="class in org.jgap">Population</A>&nbsp;a_chromosomes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Calculates and sets the fitness values on each of the given
 Chromosomes via their setFitnessValue() method.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;double</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../../org/jgap/impl/BulkFitnessOffsetRemover.html#getAbsoluteFitness(org.jgap.IChromosome)">getAbsoluteFitness</A></B>(<A HREF="../../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A>&nbsp;a_individuum)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
 Using this instance to remove the fitness offset in the
 populations brings the advantage of getting a selection
 more sensitive to the differences of fitness of the chromosomes.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="BulkFitnessOffsetRemover(org.jgap.FitnessFunction)"><!-- --></A><H3>
BulkFitnessOffsetRemover</H3>
<PRE>
public <B>BulkFitnessOffsetRemover</B>(<A HREF="../../../org/jgap/FitnessFunction.html" title="class in org.jgap">FitnessFunction</A>&nbsp;a_ff)</PRE>
<DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="evaluate(org.jgap.Population)"><!-- --></A><H3>
evaluate</H3>
<PRE>
public void <B>evaluate</B>(<A HREF="../../../org/jgap/Population.html" title="class in org.jgap">Population</A>&nbsp;a_chromosomes)</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../../org/jgap/BulkFitnessFunction.html#evaluate(org.jgap.Population)">BulkFitnessFunction</A></CODE></B></DD>
<DD>Calculates and sets the fitness values on each of the given
 Chromosomes via their setFitnessValue() method.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/BulkFitnessFunction.html#evaluate(org.jgap.Population)">evaluate</A></CODE> in class <CODE><A HREF="../../../org/jgap/BulkFitnessFunction.html" title="class in org.jgap">BulkFitnessFunction</A></CODE></DL>
</DD>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_chromosomes</CODE> - list of Chromosomes for which the fitness values
 must be computed and set</DL>
</DD>
</DL>
<HR>

<A NAME="getAbsoluteFitness(org.jgap.IChromosome)"><!-- --></A><H3>
getAbsoluteFitness</H3>
<PRE>
public double <B>getAbsoluteFitness</B>(<A HREF="../../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A>&nbsp;a_individuum)</PRE>
<DL>
<DD><p>
 Using this instance to remove the fitness offset in the
 populations brings the advantage of getting a selection
 more sensitive to the differences of fitness of the chromosomes.
 </p>
 <p>
 The disadvantage is, that the fitness values are modified.
 The modification is good for jgap's selection method but
 bad for the guys that want to see the success of your
 work, or need a proof that a GA improves over time:
 <br>
 The value of <A HREF="../../../org/jgap/Genotype.html#getFittestChromosome()"><CODE>Genotype.getFittestChromosome()</CODE></A>
 does not seem to increase over the generations. Most often
 it becomes worse. This is caused by the fact, that all
 Chromosomes are getting better over time
 (the fitness interval of all Chromosomes gets narrower) and
 the offset that may be cut becomes bigger.
 </p>
 <p>
 If you want to get an absolute value independant from the
 offset that is cut off from the chromosome's fitness value,
 this method has to be used.
 </p>
 <p>
 Stop reading here because a
 </p>
 <h4>Mathematical Proof</h4>
 <p>
 is following.
 How can it work to get the absolute value for all
 Chromosomes fitness values? Some Chromosomes may have lived
 for many generations and everytime their fitness was
 evaluated here, the old offset was added and a new one was
 calculated and subtracted from the fitness value.
 </p>
 <p>
 Each bulk fitness evaluation a Chromosome experiences,
 it's fitness value <i>F</i> get's an addition of the old offset
 <i>O<sub>(n-1)</sub></i>
 and a substraction by the new offset <i>O<sub>n</sub></i>.<br>
 <i><sub>n</sub></i> is the generation index.

 <pre>
 F<sub>1</sub> = F<sub>0</sub> + O<sub>0</sub> - O<sub>1</sub>
 F<sub>2</sub> = F<sub>1</sub> + O<sub>1</sub> - O<sub>2</sub>
 F<sub>3</sub> = F<sub>2</sub> + O<sub>2</sub> - O<sub>3</sub>

 =>

 1) F<sub>n</sub> = <b>F<sub>(n-1)</sub></b>
 + O<sub>(n-1)</sub> - O<sub>n</sub>

 2) <b>F<sub>(n-1)</sub></b> = F<sub>(n-2)</sub>
 + O<sub>(n-2)</sub> - O<sub>(n-1)</sub>

 2 in 1)
    F<sub>n</sub> = (F<sub>(n-2)</sub> + O<sub>(n-2)</sub>
 - O<sub>(n-1)</sub>) + O<sub>(n-1)</sub> - O<sub>n</sub>
    F<sub>n</sub> = F<sub>(n-2)</sub> + O<sub>(n-2)</sub> - O<sub>n</sub>

 We made a step over 2 generations: With the current offset and the
 fitness & offset of the
 "preprevious" generation we can calculate the current fitness.
 We can assume that this generation stepping works for farer steps
 <sub>m</sub> (just continue step 2) until you have a generation step value
 high enough ;-))

 => F<sub>n</sub> = F<sub>(n-m)</sub> + O<sub>(n-m)</sub> - O<sub>n</sub>

 We want to get the original absolute value of fitness:

 3) m := n

 => F<sub>n</sub> = F<sub>0</sub> + O<sub>0</sub> - O<sub>n</sub>

 solved to F<sub>0</sub> our original value:

 F<sub>0</sub> = F<sub>n</sub> + O<sub>n</sub> - O<sub>0</sub>

 And our initial offset <A HREF="../../../org/jgap/impl/BulkFitnessOffsetRemover.html#m_previousOffset"><CODE>O<sub>0</sub></CODE></A> is zero!
 </pre>
 </p>
 <p>
 This shows, that it is possible to compute the original fitness value of a
 Chromosome from it's current fitness value and the
 <A HREF="../../../org/jgap/impl/BulkFitnessOffsetRemover.html#m_previousOffset"><CODE>previous offset</CODE></A>
 regardless of the amounts of generations between original evaluation and
 the current generation.
 </p>
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_individuum</CODE> - any Chromosome that is normally being evaluated by
 this <tt>BulkFitnessFunction</tt>
<DT><B>Returns:</B><DD>the original fitness value as returned by the registered
 <A HREF="../../../org/jgap/impl/BulkFitnessOffsetRemover.html#m_ff"><CODE>fitnessFunction</CODE></A> instance.</DL>
</DD>
</DL>
<HR>

<A NAME="clone()"><!-- --></A><H3>
clone</H3>
<PRE>
public java.lang.Object <B>clone</B>()</PRE>
<DL>
<DD><B>Description copied from class: <CODE><A HREF="../../../org/jgap/BulkFitnessFunction.html#clone()">BulkFitnessFunction</A></CODE></B></DD>
<DD>Override in your implementation if necessary.
<P>
<DD><DL>
<DT><B>Specified by:</B><DD><CODE><A HREF="../../../org/jgap/util/ICloneable.html#clone()">clone</A></CODE> in interface <CODE><A HREF="../../../org/jgap/util/ICloneable.html" title="interface in org.jgap.util">ICloneable</A></CODE><DT><B>Overrides:</B><DD><CODE><A HREF="../../../org/jgap/BulkFitnessFunction.html#clone()">clone</A></CODE> in class <CODE><A HREF="../../../org/jgap/BulkFitnessFunction.html" title="class in org.jgap">BulkFitnessFunction</A></CODE></DL>
</DD>
<DD><DL>

<DT><B>Returns:</B><DD>deep clone of current instance<DT><B>Since:</B></DT>
  <DD>3.2</DD>
</DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
JGAP</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../../org/jgap/impl/BooleanGene.html" title="class in org.jgap.impl"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../../org/jgap/impl/CauchyRandomGenerator.html" title="class in org.jgap.impl"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../../index.html?org/jgap/impl/BulkFitnessOffsetRemover.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="BulkFitnessOffsetRemover.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;FIELD&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
