<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.6.0_02) on Mon Nov 05 17:25:06 CET 2007 -->
<TITLE>
AbstractSupergeneTest
</TITLE>

<META NAME="date" CONTENT="2007-11-05">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    if (location.href.indexOf('is-external=true') == -1) {
        parent.document.title="AbstractSupergeneTest";
    }
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">
<HR>


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
JGAP</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../examples/supergene/AbstractChangeFitnessFunction.html" title="class in examples.supergene"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../examples/supergene/Force.html" title="class in examples.supergene"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?examples/supergene/AbstractSupergeneTest.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="AbstractSupergeneTest.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
examples.supergene</FONT>
<BR>
Class AbstractSupergeneTest</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../resources/inherit.gif" ALT="extended by "><B>examples.supergene.AbstractSupergeneTest</B>
</PRE>
<DL>
<DT><B>Direct Known Subclasses:</B> <DD><A HREF="../../examples/supergene/SupergeneSample.html" title="class in examples.supergene">SupergeneSample</A></DD>
</DL>
<HR>
<DL>
<DT><PRE>public abstract class <B>AbstractSupergeneTest</B><DT>extends java.lang.Object</DL>
</PRE>

<P>
Abstract class for testing Supergene performance.
<P>

<P>
<DL>
<DT><B>Since:</B></DT>
  <DD>2.0</DD>
</DL>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../examples/supergene/AbstractSupergeneTest.html#DIMES">DIMES</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gene index for the dimes gene</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../examples/supergene/AbstractSupergeneTest.html#EXISTING_SOLUTIONS_ONLY">EXISTING_SOLUTIONS_ONLY</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;If set to true (required for strict tests), only tasks with existing
 solutions will be submitted as a test tasks.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../examples/supergene/AbstractSupergeneTest.html#MAX_ALLOWED_EVOLUTIONS">MAX_ALLOWED_EVOLUTIONS</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The total number of times we'll let the population evolve.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../examples/supergene/AbstractSupergeneTest.html#NICKELS">NICKELS</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gene index for the nickels gene
 Only used in the alternative presentation</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../examples/supergene/AbstractSupergeneTest.html#PENNIES">PENNIES</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gene index for the pennies gene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../examples/supergene/AbstractSupergeneTest.html#POPULATION_SIZE">POPULATION_SIZE</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Chromosome size.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../examples/supergene/AbstractSupergeneTest.html#QUARTERS">QUARTERS</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gene index for the quarters gene.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;boolean</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../examples/supergene/AbstractSupergeneTest.html#REPORT_ENABLED">REPORT_ENABLED</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../examples/supergene/AbstractSupergeneTest.html#AbstractSupergeneTest()">AbstractSupergeneTest</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>static&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../examples/supergene/AbstractSupergeneTest.html#amountOfChange(int, int, int, int)">amountOfChange</A></B>(int&nbsp;a_numQuarters,
               int&nbsp;a_numDimes,
               int&nbsp;a_numNickels,
               int&nbsp;a_numPennies)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Compute the money value from the coin information.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../examples/supergene/AbstractSupergeneTest.html#getDimesGene(org.jgap.Configuration)">getDimesGene</A></B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_conf)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../examples/supergene/AbstractSupergeneTest.html#getNickelsGene(org.jgap.Configuration)">getNickelsGene</A></B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_conf)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../examples/supergene/AbstractSupergeneTest.html#getPenniesGene(org.jgap.Configuration)">getPenniesGene</A></B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_conf)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;<A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../examples/supergene/AbstractSupergeneTest.html#getQuartersGene(org.jgap.Configuration)">getQuartersGene</A></B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_conf)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>abstract &nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../examples/supergene/AbstractSupergeneTest.html#makeChangeForAmount(int)">makeChangeForAmount</A></B>(int&nbsp;a_targetChangeAmount)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Executes the genetic algorithm to determine the minimum number of
 coins necessary to make up the given target amount of change.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;<A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A></CODE></FONT></TD>
<TD><CODE><B><A HREF="../../examples/supergene/AbstractSupergeneTest.html#report(examples.supergene.SupergeneChangeFitnessFunction, org.jgap.Genotype)">report</A></B>(<A HREF="../../examples/supergene/SupergeneChangeFitnessFunction.html" title="class in examples.supergene">SupergeneChangeFitnessFunction</A>&nbsp;a_fitnessFunction,
       <A HREF="../../org/jgap/Genotype.html" title="class in org.jgap">Genotype</A>&nbsp;a_population)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Write report on eveluation to the given stream.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../examples/supergene/AbstractSupergeneTest.html#solve(org.jgap.Configuration, int, examples.supergene.SupergeneChangeFitnessFunction, org.jgap.Gene[])">solve</A></B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_conf,
      int&nbsp;a_targetChangeAmount,
      <A HREF="../../examples/supergene/SupergeneChangeFitnessFunction.html" title="class in examples.supergene">SupergeneChangeFitnessFunction</A>&nbsp;a_fitnessFunction,
      <A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>[]&nbsp;a_sampleGenes)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Find and print the solution, return the solution error.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../examples/supergene/AbstractSupergeneTest.html#test()">test</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Test the method, returns the sum of all differences between
 the required and obtained excange amount.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="DIMES"><!-- --></A><H3>
DIMES</H3>
<PRE>
public static final int <B>DIMES</B></PRE>
<DL>
<DD>Gene index for the dimes gene
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#examples.supergene.AbstractSupergeneTest.DIMES">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="QUARTERS"><!-- --></A><H3>
QUARTERS</H3>
<PRE>
public static final int <B>QUARTERS</B></PRE>
<DL>
<DD>Gene index for the quarters gene.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#examples.supergene.AbstractSupergeneTest.QUARTERS">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="NICKELS"><!-- --></A><H3>
NICKELS</H3>
<PRE>
public static final int <B>NICKELS</B></PRE>
<DL>
<DD>Gene index for the nickels gene
 Only used in the alternative presentation
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#examples.supergene.AbstractSupergeneTest.NICKELS">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="PENNIES"><!-- --></A><H3>
PENNIES</H3>
<PRE>
public static final int <B>PENNIES</B></PRE>
<DL>
<DD>Gene index for the pennies gene.
 Only used in the alternative presentation
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#examples.supergene.AbstractSupergeneTest.PENNIES">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="MAX_ALLOWED_EVOLUTIONS"><!-- --></A><H3>
MAX_ALLOWED_EVOLUTIONS</H3>
<PRE>
public static int <B>MAX_ALLOWED_EVOLUTIONS</B></PRE>
<DL>
<DD>The total number of times we'll let the population evolve.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="POPULATION_SIZE"><!-- --></A><H3>
POPULATION_SIZE</H3>
<PRE>
public static int <B>POPULATION_SIZE</B></PRE>
<DL>
<DD>Chromosome size.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="REPORT_ENABLED"><!-- --></A><H3>
REPORT_ENABLED</H3>
<PRE>
public static boolean <B>REPORT_ENABLED</B></PRE>
<DL>
<DL>
</DL>
</DL>
<HR>

<A NAME="EXISTING_SOLUTIONS_ONLY"><!-- --></A><H3>
EXISTING_SOLUTIONS_ONLY</H3>
<PRE>
public static boolean <B>EXISTING_SOLUTIONS_ONLY</B></PRE>
<DL>
<DD>If set to true (required for strict tests), only tasks with existing
 solutions will be submitted as a test tasks.
<P>
<DL>
</DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="AbstractSupergeneTest()"><!-- --></A><H3>
AbstractSupergeneTest</H3>
<PRE>
public <B>AbstractSupergeneTest</B>()</PRE>
<DL>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="getDimesGene(org.jgap.Configuration)"><!-- --></A><H3>
getDimesGene</H3>
<PRE>
protected <A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A> <B>getDimesGene</B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_conf)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_conf</CODE> - the configuration to use
<DT><B>Returns:</B><DD>created Dimes gene instance</DL>
</DD>
</DL>
<HR>

<A NAME="getNickelsGene(org.jgap.Configuration)"><!-- --></A><H3>
getNickelsGene</H3>
<PRE>
protected <A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A> <B>getNickelsGene</B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_conf)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_conf</CODE> - the configuration to use
<DT><B>Returns:</B><DD>created Nickels gene instance</DL>
</DD>
</DL>
<HR>

<A NAME="getPenniesGene(org.jgap.Configuration)"><!-- --></A><H3>
getPenniesGene</H3>
<PRE>
protected <A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A> <B>getPenniesGene</B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_conf)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_conf</CODE> - the configuration to use
<DT><B>Returns:</B><DD>created Pennies (1) gene instance</DL>
</DD>
</DL>
<HR>

<A NAME="getQuartersGene(org.jgap.Configuration)"><!-- --></A><H3>
getQuartersGene</H3>
<PRE>
protected <A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A> <B>getQuartersGene</B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_conf)</PRE>
<DL>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_conf</CODE> - the configuration to use
<DT><B>Returns:</B><DD>created Quarters gene instance</DL>
</DD>
</DL>
<HR>

<A NAME="amountOfChange(int, int, int, int)"><!-- --></A><H3>
amountOfChange</H3>
<PRE>
public static int <B>amountOfChange</B>(int&nbsp;a_numQuarters,
                                 int&nbsp;a_numDimes,
                                 int&nbsp;a_numNickels,
                                 int&nbsp;a_numPennies)</PRE>
<DL>
<DD>Compute the money value from the coin information.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="makeChangeForAmount(int)"><!-- --></A><H3>
makeChangeForAmount</H3>
<PRE>
public abstract int <B>makeChangeForAmount</B>(int&nbsp;a_targetChangeAmount)
                                 throws java.lang.Exception</PRE>
<DL>
<DD>Executes the genetic algorithm to determine the minimum number of
 coins necessary to make up the given target amount of change. The
 solution will then be written to System.out.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_targetChangeAmount</CODE> - the target amount of change for which this
 method is attempting to produce the minimum number of coins
<DT><B>Returns:</B><DD>absolute difference between the required and computed change
 amount
<DT><B>Throws:</B>
<DD><CODE>java.lang.Exception</CODE></DL>
</DD>
</DL>
<HR>

<A NAME="report(examples.supergene.SupergeneChangeFitnessFunction, org.jgap.Genotype)"><!-- --></A><H3>
report</H3>
<PRE>
public <A HREF="../../org/jgap/IChromosome.html" title="interface in org.jgap">IChromosome</A> <B>report</B>(<A HREF="../../examples/supergene/SupergeneChangeFitnessFunction.html" title="class in examples.supergene">SupergeneChangeFitnessFunction</A>&nbsp;a_fitnessFunction,
                          <A HREF="../../org/jgap/Genotype.html" title="class in org.jgap">Genotype</A>&nbsp;a_population)</PRE>
<DL>
<DD>Write report on eveluation to the given stream.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_fitnessFunction</CODE> - p_SupergeneChangeFitnessFunction<DD><CODE>a_population</CODE> - Genotype
<DT><B>Returns:</B><DD>Chromosome</DL>
</DD>
</DL>
<HR>

<A NAME="test()"><!-- --></A><H3>
test</H3>
<PRE>
public int <B>test</B>()</PRE>
<DL>
<DD>Test the method, returns the sum of all differences between
 the required and obtained excange amount. One exception counts
 as 1000 on the error score.
<P>
<DD><DL>
</DL>
</DD>
</DL>
<HR>

<A NAME="solve(org.jgap.Configuration, int, examples.supergene.SupergeneChangeFitnessFunction, org.jgap.Gene[])"><!-- --></A><H3>
solve</H3>
<PRE>
protected int <B>solve</B>(<A HREF="../../org/jgap/Configuration.html" title="class in org.jgap">Configuration</A>&nbsp;a_conf,
                    int&nbsp;a_targetChangeAmount,
                    <A HREF="../../examples/supergene/SupergeneChangeFitnessFunction.html" title="class in examples.supergene">SupergeneChangeFitnessFunction</A>&nbsp;a_fitnessFunction,
                    <A HREF="../../org/jgap/Gene.html" title="interface in org.jgap">Gene</A>[]&nbsp;a_sampleGenes)
             throws <A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></PRE>
<DL>
<DD>Find and print the solution, return the solution error.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>a_conf</CODE> - the configuration to use
<DT><B>Returns:</B><DD>absolute difference between the required and computed change
<DT><B>Throws:</B>
<DD><CODE><A HREF="../../org/jgap/InvalidConfigurationException.html" title="class in org.jgap">InvalidConfigurationException</A></CODE></DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-all.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
JGAP</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;<A HREF="../../examples/supergene/AbstractChangeFitnessFunction.html" title="class in examples.supergene"><B>PREV CLASS</B></A>&nbsp;
&nbsp;<A HREF="../../examples/supergene/Force.html" title="class in examples.supergene"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?examples/supergene/AbstractSupergeneTest.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="AbstractSupergeneTest.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
